{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "%config IPCompleter.greedy=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import f1_score, accuracy_score\n",
    "from scipy.special import softmax\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import xgboost\n",
    "import optuna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pandas in practice - Neurofirbromatosis detection\n",
    "\n",
    "Welcome to the lecture about practical use cases of Pandas library! In this notebook Pandas library was used for ETA and data formatting in real-life research project: usage of Artificial Intelligence for Neurofibromatosis diagnosis based on human genome. Our goal is to develop artificial intelligence system able to classify sick people among healthy ones by interpreting their genetic code. Let's get started!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Contents of this notebook\n",
    "\n",
    "- Pandas: EDA (*Exploratory Data Analysis*)\n",
    "- Pandas: Data I/O, preprocessing and formatting\n",
    "- Sklearn: Neurofibromatosis detection\n",
    "- Oputna: automatic hyperparameters tuning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libraries required:\n",
    "\n",
    "- pandas\n",
    "- numpy\n",
    "- sklearn\n",
    "- matplotlib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test data creation\n",
    "\n",
    "Unfortunately, this notebook cannot use original, real data because of protection of sensitive informations and copyrights. So, script below generates dummy csv files which resemble true data. This operation can be skipped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_avaliable_genes(number_of_possibilities: int = 1500):\n",
    "    chromosomes = [\"chr\" + str(i) for i in range(1, 23)]\n",
    "    chromosomes.extend([\"chrX\", \"chrY\"])\n",
    "\n",
    "    positions = list(range(1,2000))\n",
    "\n",
    "    mutations = []\n",
    "    nucleotydes = [\"A\", \"C\", \"T\", \"G\"]\n",
    "    for i in nucleotydes:\n",
    "        for j in nucleotydes:\n",
    "            mutations.append(f\"{i}/{j}\")\n",
    "\n",
    "    chromosomes = pd.Series(chromosomes)\n",
    "    positions = pd.Series(positions)\n",
    "    mutations = pd.Series(mutations)    \n",
    "\n",
    "\n",
    "\n",
    "    chromosome_choices = np.random.randint(low=0, high=chromosomes.shape, size=number_of_possibilities)\n",
    "    picked_chromosomes = pd.Series(data=chromosomes.iloc[chromosome_choices])\n",
    "    picked_chromosomes.index = list(range(picked_chromosomes.shape[0]))\n",
    "\n",
    "    position_choices = np.random.randint(low=0, high=positions.shape, size=number_of_possibilities)\n",
    "    picked_positions = positions.iloc[position_choices]\n",
    "    picked_positions.index = list(range(picked_positions.shape[0]))\n",
    "\n",
    "    mutation_choices = np.random.randint(low=0, high=mutations.shape, size=number_of_possibilities)\n",
    "    picked_mutations = mutations.iloc[mutation_choices]\n",
    "    picked_mutations.index = list(range(picked_mutations.shape[0]))\n",
    "\n",
    "    options = {\n",
    "        \"chromosome\": picked_chromosomes,\n",
    "        \"position\": picked_positions,\n",
    "        \"mutation\": picked_mutations\n",
    "    }\n",
    "\n",
    "    avaliable_options = pd.DataFrame(options)\n",
    "    return avaliable_options\n",
    "\n",
    "def get_indicies(amount: int, mean: int, std: int, max_value: int):\n",
    "    indicies = np.random.randn(amount) * mean + std\n",
    "    indicies = np.array(indicies, dtype=np.int64)\n",
    "    indicies = indicies[indicies >= 0]\n",
    "    indicies = indicies[indicies < max_value]\n",
    "    return np.unique(indicies)\n",
    "\n",
    "def create_file_frame(is_healthy: bool, avaliable_options: pd.DataFrame):\n",
    "    amount = 1000\n",
    "    mean = int(avaliable_options.shape[0] * 0.5) - 100 if is_healthy else int(avaliable_options.shape[0] * 0.5) + 100\n",
    "    std = 300 if is_healthy else 600\n",
    "    indicies = get_indicies(amount, mean, std, avaliable_options.shape[0])\n",
    "    frame = avaliable_options.iloc[indicies]\n",
    "    frame.index = list(range(frame.shape[0]))\n",
    "    \n",
    "    chr_name = \"chr\" if is_healthy else \"CHROMOSOME\"\n",
    "    pos_name = \"pos\" if is_healthy else \"POSITION\"\n",
    "    mut_name = \"mut\" if is_healthy else \"MUTATION\"\n",
    "    \n",
    "    frame = frame.rename(columns={\"chromosome\": chr_name, \"position\": \"pos\", \"mutation\": \"mut\"})\n",
    "    \n",
    "    if is_healthy:\n",
    "        frame[\"num_pl_a\"] = np.nan\n",
    "    else:\n",
    "        frame[\"is_ok\"] = True\n",
    "        frame[\"approved_ratio\"] = 0.9 \n",
    "    \n",
    "    return frame\n",
    "\n",
    "def create_files(number_of_possibilities: int, healthy: int, sick: int, healthy_folder_path: str, sick_folder_path: str):\n",
    "    options = get_avaliable_genes()\n",
    "    for i in range(healthy):\n",
    "        frame = create_file_frame(True, options)\n",
    "        frame.to_csv(os.path.join(healthy_folder_path, f\"{i}.csv\"))\n",
    "    for i in range(sick):\n",
    "        frame = create_file_frame(False, options)\n",
    "        frame.to_csv(os.path.join(sick_folder_path, f\"{i}.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_files(1500, 60, 42, \"healthy\", \"sick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Exploratory Data Analysis (*EDA*)\n",
    "\n",
    "In this segment we will perform data checks to figure out what type of data we got"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_healthy = \"healthy\"\n",
    "path_to_sick = \"sick\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load files\n",
    "\n",
    "First, we will load csv files representing people suffering from Neurofirbomatosis disease and healthy ones. We will pandas method `read_csv` to create person's dataframe from file. Pandas is able to read alot of different file types, but for the sake of this example we will use csv files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "healthy_people_file_paths = [os.path.join(path_to_healthy, f) for f in os.listdir(path_to_healthy)]\n",
    "sick_people_file_paths = [os.path.join(path_to_sick,f) for f in os.listdir(path_to_sick)]\n",
    "\n",
    "df_healthy = [pd.read_csv(f) for f in healthy_people_file_paths]\n",
    "df_sick = [pd.read_csv(f) for f in sick_people_file_paths]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check number of files\n",
    "\n",
    "Are your classes (healthy, sick) evenly ballanced? It is important to know, because classifiers need to be specially tuned in case of class inbalance. It is a simple check, but it is very effective!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of healthy people: 60\n",
      "Number of sick people: 42\n"
     ]
    }
   ],
   "source": [
    "print(f\"Number of healthy people: {len(df_healthy)}\")\n",
    "print(f\"Number of sick people: {len(df_sick)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Turns out that sick people consist only 41% of the dataset. It is not a drastic class inbalance and classifiers should handle it without additional tuning. Nonetheless, it is important to be aware of it!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick peek into the dataframes\n",
    "\n",
    "We will use two methods: `head` and `tail`. `head` method shows first 5 rows of the dataframe when `tail` shows last 5 rows (methods can be provided with number of rows to be selected). This peek will allow us to quickly build an intuition about data we work with. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>num_pl_a</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr19</td>\n",
       "      <td>29</td>\n",
       "      <td>A/A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr15</td>\n",
       "      <td>1760</td>\n",
       "      <td>T/G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr14</td>\n",
       "      <td>1032</td>\n",
       "      <td>C/T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr10</td>\n",
       "      <td>1253</td>\n",
       "      <td>T/C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr11</td>\n",
       "      <td>1748</td>\n",
       "      <td>C/A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0    chr   pos  mut  num_pl_a\n",
       "0           0  chr19    29  A/A       NaN\n",
       "1           1  chr15  1760  T/G       NaN\n",
       "2           2  chr14  1032  C/T       NaN\n",
       "3           3  chr10  1253  T/C       NaN\n",
       "4           4  chr11  1748  C/A       NaN"
      ]
     },
     "execution_count": 208,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_healthy[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>is_ok</th>\n",
       "      <th>approved_ratio</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>477</td>\n",
       "      <td>chr22</td>\n",
       "      <td>1419</td>\n",
       "      <td>T/A</td>\n",
       "      <td>True</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>478</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1612</td>\n",
       "      <td>A/A</td>\n",
       "      <td>True</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>479</td>\n",
       "      <td>chr11</td>\n",
       "      <td>1987</td>\n",
       "      <td>T/T</td>\n",
       "      <td>True</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>480</td>\n",
       "      <td>chr15</td>\n",
       "      <td>68</td>\n",
       "      <td>G/G</td>\n",
       "      <td>True</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>481</td>\n",
       "      <td>chrX</td>\n",
       "      <td>364</td>\n",
       "      <td>C/C</td>\n",
       "      <td>True</td>\n",
       "      <td>0.9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Unnamed: 0 CHROMOSOME   pos  mut  is_ok  approved_ratio\n",
       "477         477      chr22  1419  T/A   True             0.9\n",
       "478         478       chr6  1612  A/A   True             0.9\n",
       "479         479      chr11  1987  T/T   True             0.9\n",
       "480         480      chr15    68  G/G   True             0.9\n",
       "481         481       chrX   364  C/C   True             0.9"
      ]
     },
     "execution_count": 209,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sick[0].tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Something interesting! It seems that healthy and sick persons csv files contains different columns! Let's take a look on it in more focused manner."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check columns\n",
    "\n",
    "First, let's check what columns are in every file and group. To do this we will use dataframe fields `columns` and `dtypes`. Also, we will use `groupby` function to get unique columns and `describe` to summarize results of our operations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_column_types(df_list: list) -> pd.DataFrame:\n",
    "    d = {\n",
    "        \"column\": [],\n",
    "        \"type\":[]\n",
    "    }\n",
    "    for df in df_list:\n",
    "        for i in range(len(df.columns)):\n",
    "            d[\"column\"].append(df.columns[i])\n",
    "            d[\"type\"].append(df.dtypes[i])\n",
    "    \n",
    "    df = pd.DataFrame(d)\n",
    "    df = df.groupby(\"column\")\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>column</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>int64</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>object</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mut</th>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>object</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>num_pl_a</th>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>float64</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <td>60</td>\n",
       "      <td>1</td>\n",
       "      <td>int64</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            type                     \n",
       "           count unique      top freq\n",
       "column                               \n",
       "Unnamed: 0    60      1    int64   60\n",
       "chr           60      1   object   60\n",
       "mut           60      1   object   60\n",
       "num_pl_a      60      1  float64   60\n",
       "pos           60      1    int64   60"
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "healthy_person_columns = check_column_types(df_healthy)\n",
    "healthy_person_columns.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"4\" halign=\"left\">type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>unique</th>\n",
       "      <th>top</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>column</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>object</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>int64</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>approved_ratio</th>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>float64</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_ok</th>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>bool</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mut</th>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>object</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <td>42</td>\n",
       "      <td>1</td>\n",
       "      <td>int64</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                type                     \n",
       "               count unique      top freq\n",
       "column                                   \n",
       "CHROMOSOME        42      1   object   42\n",
       "Unnamed: 0        42      1    int64   42\n",
       "approved_ratio    42      1  float64   42\n",
       "is_ok             42      1     bool   42\n",
       "mut               42      1   object   42\n",
       "pos               42      1    int64   42"
      ]
     },
     "execution_count": 212,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sick_person_columns = check_column_types(df_sick)\n",
    "sick_person_columns.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok, so what do we got?\n",
    "\n",
    "Healthy persons csvs contain classes: \"Unnamed: 0\", \"chr\", \"mut\", \"num_pl_a\", \"pos\" and sick persons csvs contain classes:  \"CHROMOSOME\", \"Unnamed: 0\", \"approve_ratio\", \"is_ok\", \"mut\", \"pos\".\n",
    "\n",
    "It means that these csv files are of different format and we need more details to know how to combine those two groups. We also know, what type of fileds are stored in these columns (strings are considered `object`s in pandas) and that in each column there is only one datatype (columns `count` and `type freq` are equal to eachother)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Time to ask questions to person, who knows context of this data\n",
    "\n",
    "As a Data Scientists, we know how to operate on data, but we often don't know the context behind the data. These csv files has been given to us by scientist in the field of Biology. She knows what each column really means, so it is time to ask her a few questions:\n",
    "\n",
    "- Are columns `chr` and `CHROMOSOME` equivalents of each other?\n",
    "- What columns are essential to encode genetic mutation?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Getting ansers back\n",
    "\n",
    "- Yes, columns `chr` and CHROMOSOME are equivalents of each other\n",
    "- We need 'chr', 'pos' and 'mut'\n",
    "\n",
    "Good. We can proceed then!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unify the DataFrames: Filter out unnecessary rows\n",
    "\n",
    "The intersection of sets of columns of healthy and sick persons are {'chr' <-> 'CHROMOSOME', 'pos', 'mut'}. It seems like for now no other column is needed, so let's take them.\n",
    "\n",
    "To do that we will use operation: `df[[\"first_column_to_take_name', \"second_column_to_take_name\"...]]`. It allows to take from DataFrame only these columns, that we specify. If we are to take only one column, then resulting object would be type `pd.Series`, but in this example we will use multiple columns, so it will be of type `pd.DataFrame`.\n",
    "\n",
    "Let's filter out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_healthy_filtered = [df[[\"chr\", \"pos\", \"mut\"]] for df in df_healthy]\n",
    "df_sick_filtered = [df[[\"CHROMOSOME\", \"pos\", \"mut\"]] for df in df_sick]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick check, if that worked:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>29</td>\n",
       "      <td>A/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>1760</td>\n",
       "      <td>T/G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr14</td>\n",
       "      <td>1032</td>\n",
       "      <td>C/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr10</td>\n",
       "      <td>1253</td>\n",
       "      <td>T/C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr11</td>\n",
       "      <td>1748</td>\n",
       "      <td>C/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   pos  mut\n",
       "0  chr19    29  A/A\n",
       "1  chr15  1760  T/G\n",
       "2  chr14  1032  C/T\n",
       "3  chr10  1253  T/C\n",
       "4  chr11  1748  C/A"
      ]
     },
     "execution_count": 214,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_healthy_filtered[0].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROMOSOME</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>1597</td>\n",
       "      <td>G/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>29</td>\n",
       "      <td>A/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>1960</td>\n",
       "      <td>G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>514</td>\n",
       "      <td>C/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>1792</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROMOSOME   pos  mut\n",
       "0      chr18  1597  G/A\n",
       "1      chr19    29  A/A\n",
       "2       chr7  1960  G/T\n",
       "3      chr13   514  C/A\n",
       "4      chr22  1792  G/G"
      ]
     },
     "execution_count": 215,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sick_filtered[0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unify the DataFrames: rename columns\n",
    "\n",
    "There is one column that encodes the same data, yet it has different name in each of sets: \"chr\" in healthy persons group and \"CHROMOSOME\" in sick ones. That is not good! Because of that, we can't merge these dataframes together, perform any complicated operation and we have to constantly remember where each version of column goes. \n",
    "\n",
    "Solution is simple: rename the column in one set.\n",
    "\n",
    "Let's rename \"CHROMOSOME\" in sick person gropu to \"chr\". To do that we will use function `rename`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_sick_filtered = [df.rename(columns={\"CHROMOSOME\": \"chr\"}) for df in df_sick_filtered]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see if that worked:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>1597</td>\n",
       "      <td>G/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>29</td>\n",
       "      <td>A/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr7</td>\n",
       "      <td>1960</td>\n",
       "      <td>G/T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr13</td>\n",
       "      <td>514</td>\n",
       "      <td>C/A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr22</td>\n",
       "      <td>1792</td>\n",
       "      <td>G/G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr   pos  mut\n",
       "0  chr18  1597  G/A\n",
       "1  chr19    29  A/A\n",
       "2   chr7  1960  G/T\n",
       "3  chr13   514  C/A\n",
       "4  chr22  1792  G/G"
      ]
     },
     "execution_count": 217,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sick_filtered[0].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check how much rows are there in each file\n",
    "\n",
    "We know a thing or two regarding files columns, but what about rows count? Let's find out!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 218,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rows_statistics(dfs: list) -> None:\n",
    "    counts = [df.shape[0] for df in dfs]\n",
    "    counts = np.array(counts)\n",
    "    \n",
    "    return {\n",
    "        \"sum\": counts.sum(),\n",
    "        \"mean\": counts.mean(),\n",
    "        \"min\": counts.min(),\n",
    "        \"max\": counts.max(),\n",
    "        \"std\": counts.std()\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sum': 30495,\n",
       " 'mean': 508.25,\n",
       " 'min': 481,\n",
       " 'max': 539,\n",
       " 'std': 13.796889746122735}"
      ]
     },
     "execution_count": 219,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_rows_statistics(df_healthy_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sum': 21187,\n",
       " 'mean': 504.45238095238096,\n",
       " 'min': 481,\n",
       " 'max': 532,\n",
       " 'std': 11.613215749539465}"
      ]
     },
     "execution_count": 220,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_rows_statistics(df_sick_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Merge dataframes in one big dataframe\n",
    "\n",
    "Now, when we have uniformed dataframes we want to make some statistics regarding values stored in `chr`, `pos` and `mut` columns. To do it, first we need to merge all these 100 small dataframes in one, big dataframe.\n",
    "\n",
    "To do that we will use function `concat` with parameter `axis` equal to 0. Function `concat` concatenates one dataframe to another. `axis` parameter determines which axes will be concatenated. 0 means \"concatenate rows of the second dataframe below rows of the first dataframe\" when 1 means \"concatenate columns of the second dataframe after columns of the first dataframe\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hs = pd.concat([*df_healthy_filtered, *df_sick_filtered], axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's see if that worked:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>29</td>\n",
       "      <td>A/A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     chr  pos  mut\n",
       "0  chr19   29  A/A"
      ]
     },
     "execution_count": 222,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hs.head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(51682, 3)"
      ]
     },
     "execution_count": 223,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51682"
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array([df.shape[0] for df in [*df_healthy_filtered, *df_sick_filtered]]).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Works fine! It has three columns (as we wanted) and as many rows as all the rows from all the dataframes. So it is a valid concatenation. We can proceed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get unique values for each column\n",
    "\n",
    "We have all the data in one, neat dataframe. It is time to check how many unique values are there in each column. To do it we will use `unique` funciton."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(24,)"
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_unique = df_hs[\"chr\"].unique()\n",
    "chr_unique.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1063,)"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_unique = df_hs[\"pos\"].unique()\n",
    "pos_unique.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16,)"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mut_unique = df_hs[\"mut\"].unique()\n",
    "mut_unique.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, so now we know that there are 24 different pairs of chromosomes in the set, there are 1063 unique positions and 16 different mutaiton types. That's awesome... WAIT!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Is there something wrong with the data?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yes, it is! Humans have only 23 different pairs of chromosomes! Strange... let's see what values are there:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr19', 'chr15', 'chr14', 'chr10', 'chr11', 'chr1', 'chr16',\n",
       "       'chrY', 'chr7', 'chrX', 'chr22', 'chr20', 'chr4', 'chr6', 'chr8',\n",
       "       'chr9', 'chr18', 'chr5', 'chr13', 'chr3', 'chr2', 'chr12', 'chr21',\n",
       "       'chr17'], dtype=object)"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What a mess.. let's order them alphabetically:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\n",
       "       'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21',\n",
       "       'chr22', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9',\n",
       "       'chrX', 'chrY'], dtype=object)"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sort(chr_unique)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, these files clearly have chromosomes: from chr1 to chr22 plus chrX and chrY. We found an error in the data, but what values are bad?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Time to ask questions to person, who knows context of this data\n",
    "\n",
    "Question is simple: \"what's going on with these data?\"\n",
    "Answer from the person, who knows the data: \"Hmmm... it seams that the company we hired to get that data made a blunder. Thanks! Now we know, to never hire them again! About these chromosomes... just get rid of any row that has chr=\"chr22\".\"\n",
    "\n",
    "Okay! It's time to perform filtering then!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get rows that satisfies boolean condition\n",
    "\n",
    "If pandas dataframe will get in it's square parenthesis series of boolean values (True, False, False, True...) it will return Dataframe containing rows which corresponding position of input series had True value. So, let's do the filtering!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_hs = df_hs[df_hs[\"chr\"] != \"chr22\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quick check:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(23,)"
      ]
     },
     "execution_count": 231,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_unique = df_hs[\"chr\"].unique()\n",
    "chr_unique.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['chr1', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\n",
       "       'chr16', 'chr17', 'chr18', 'chr19', 'chr2', 'chr20', 'chr21',\n",
       "       'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chrX',\n",
       "       'chrY'], dtype=object)"
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.sort(chr_unique)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Outstanding! It worked! But wait... we need to update rows statistics now!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_healthy_filtered = [df[df[\"chr\"] != \"chr22\"] for df in df_healthy_filtered]\n",
    "df_sick_filtered = [df[df[\"chr\"] != \"chr22\"] for df in df_sick_filtered]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sum': 29170,\n",
       " 'mean': 486.1666666666667,\n",
       " 'min': 460,\n",
       " 'max': 513,\n",
       " 'std': 13.309353436169951}"
      ]
     },
     "execution_count": 234,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_rows_statistics(df_healthy_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'sum': 20296,\n",
       " 'mean': 483.23809523809524,\n",
       " 'min': 462,\n",
       " 'max': 510,\n",
       " 'std': 10.960555955245725}"
      ]
     },
     "execution_count": 235,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_rows_statistics(df_sick_filtered)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get histogram of values apperance\n",
    "\n",
    "We know how many unique values are in each column. Now, we would like to know, how often each of them appears. Let's make the histograms!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD5CAYAAADLL+UrAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAezklEQVR4nO3df5RcZZ3n8ffHAAbTmh+ibSaJJg5ZPEAESQtx3cN0QENA1zAeZWFdCRxm485BR53MLOiIOBpm4wgyiMqa3WQSEG1ilElEFDORjMtBfhgRwg8xAQJJJiRqh0hIQAPf/eN5mtwUVV1V3Z0Klft5nVOnbz33+d7nuT/qe289dataEYGZmZXDKw50B8zMrHWc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErkkHoVJB0F3FAoejPwWeDaXD4R2ACcFRHbJQm4CjgD2AWcFxG/yMuaDXwmL2deRCzpr+0jjjgiJk6c2MTq7OuZZ55hxIgR+z3GcQe+rYM9rh362C5x7dDHwcQBrFmz5rcR8bqqMyOi4QcwDHgSeBPwj8DFufxi4It5+gzgh4CAacCduXwM8Gj+OzpPj+6vvalTp8Zg3HrrrS2JcdyBb+tgj2uHPrZLXDv0cTBxERHAz6NGXm12eOdU4JGIeByYBfRdqS8BzszTs4Brc9t3AKMkjQVOA1ZGRG9EbAdWAjObbN/MzAah2aR/NvDtPN0ZEVvy9JNAZ54eB2wsxGzKZbXKzcysRRQN/gyDpMOAfweOiYitkp6KiFGF+dsjYrSkm4D5EXFbLl8FXAR0A8MjYl4uvwTYHRGXV7QzB5gD0NnZObWnp2fAK7dz5046Ojr2e4zjDnxbB3tcO/SxXeLaoY+DiQOYPn36mojoqjqz1rhP5YM0bPPjwvOHgbF5eizwcJ7+BnBOZT3gHOAbhfJ96lV7eEy/vePaoY/tEtcOfWyXuHbo42DiIoZuTP8c9g7tAKwAZufp2cDyQvm5SqYBOyINA90CzJA0WtJoYEYuMzOzFql7yyaApBHAu4GPFIrnA0slXQA8DpyVy28m3cGznnTL5vkAEdEr6QvA3bne5yOid9BrYGZmDWso6UfEM8BrK8p+R7qbp7JuABfWWM4iYFHz3TQzs6Hgb+SamZWIk76ZWYk0NLxjZi+1dvMOzrv4B03HzZ2yp+m4YsyG+e9puk2zPr7SNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MSOai/nDWQL88M5IszlXH+8oyZvVz5St/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKpKGkL2mUpGWSfiXpIUnvkDRG0kpJ6/Lf0bmuJH1F0npJ90k6obCc2bn+Okmz99dKmZlZdY1e6V8F/Cgi3gIcBzwEXAysiojJwKr8HOB0YHJ+zAGuAZA0BrgUOAk4Ebi070RhZmatUTfpSxoJnAwsBIiIP0TEU8AsYEmutgQ4M0/PAq6N5A5glKSxwGnAyojojYjtwEpg5hCui5mZ1aGI6L+CdDywAHiQdJW/Bvg4sDkiRuU6ArZHxChJNwHzI+K2PG8VcBHQDQyPiHm5/BJgd0RcXtHeHNI7BDo7O6f29PQMeOW29e5g6+7mYjoPp+mYyrgp40Y2HLdz5046Ojqabq8d4tqhj4OJG8jxBQM7xnx8DV1cO/RxMHEA06dPXxMRXdXmNfLTyocAJwAfi4g7JV3F3qEcACIiJPV/9mhQRCwgnWTo6uqK7u7uAS/r6uuXc8Xa5n49eu6UPU3HVMZt+FB3w3GrV69mIOvYDnHt0MfBxA3k+IKBHWM+voYurh36OJi4ehoZ098EbIqIO/PzZaSTwNY8bEP+uy3P3wxMKMSPz2W1ys3MrEXqJv2IeBLYKOmoXHQqaahnBdB3B85sYHmeXgGcm+/imQbsiIgtwC3ADEmj8we4M3KZmZm1SKPvMT8GXC/pMOBR4HzSCWOppAuAx4Gzct2bgTOA9cCuXJeI6JX0BeDuXO/zEdE7JGthZmYNaSjpR8QvgWofCpxapW4AF9ZYziJgURP9MzOzIeRv5JqZlYiTvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYk0/x1yM7OSmHjxD16cnjtlD+cVnjdqoHGLZ45oOqYRvtI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJp6KeVJW0AngaeB/ZERJekMcANwERgA3BWRGyXJOAq4AxgF3BeRPwiL2c28Jm82HkRsWToVsUOpIlVfjq2FT9Fu2H+e5pevlmZNXOlPz0ijo+Irvz8YmBVREwGVuXnAKcDk/NjDnANQD5JXAqcBJwIXCpp9OBXwczMGjWY4Z1ZQN+V+hLgzEL5tZHcAYySNBY4DVgZEb0RsR1YCcwcRPtmZtakRpN+AD+WtEbSnFzWGRFb8vSTQGeeHgdsLMRuymW1ys3MrEUUEfUrSeMiYrOk15Ou0D8GrIiIUYU62yNitKSbgPkRcVsuXwVcBHQDwyNiXi6/BNgdEZdXtDWHNCxEZ2fn1J6engGv3LbeHWzd3VxM5+E0HVMZN2XcyIbjdu7cSUdHR9Ptvdzi1m7e8ZKyodiW9RS3dau3yUCOLxjYdin78TWUcc3EFI/rVhzPRZNGDhvQNgGYPn36msJQ/D4a+iA3Ijbnv9sk3Ugak98qaWxEbMnDN9ty9c3AhEL4+Fy2mZT4i+Wrq7S1AFgA0NXVFd3d3ZVVGnb19cu5Ym1z/wZ47pQ9TcdUxm34UHfDcatXr2Yg6/hyi6v2wetQbMt6itu61dtkIMcXDGy7lP34Gsq4ZmLOq/gfua3a35D+R+5g8l8tdYd3JI2Q9Oq+aWAGcD+wApidq80GlufpFcC5SqYBO/Iw0C3ADEmj8we4M3KZmZm1SCOnn07gxnQnJocA34qIH0m6G1gq6QLgceCsXP9m0u2a60m3bJ4PEBG9kr4A3J3rfT4ieodsTczMrK66ST8iHgWOq1L+O+DUKuUBXFhjWYuARc1308zMhoK/kWtmViJO+mZmJeKkb2ZWIk76ZmYl4qRvZlYiTvpmZiXipG9mViLNfzfYXraq/aZ9pVb8xr2ZvXz5St/MrESc9M3MSsRJ38ysRDymvx80Mrbex2PsZtZKvtI3MysRJ30zsxJx0jczKxGP6ZvZy16tz8kG8tlW2T8Pc9I3s4bVu0nBNya8/Hl4x8ysRJz0zcxKxEnfzKxEnPTNzErESd/MrEQaTvqShkm6R9JN+fkkSXdKWi/pBkmH5fJX5ufr8/yJhWV8Kpc/LOm0IV8bMzPrVzNX+h8HHio8/yJwZUQcCWwHLsjlFwDbc/mVuR6SjgbOBo4BZgJflzRscN03M7NmNHSfvqTxwHuAy4C/liTgFOC/5ipLgM8B1wCz8jTAMuCruf4soCcingMek7QeOBH42ZCsiVlJ+Af9bDAUEfUrScuA/wW8Gvgb4Dzgjnw1j6QJwA8j4lhJ9wMzI2JTnvcIcBLpRHBHRHwzly/MMcsq2poDzAHo7Oyc2tPTM+CV29a7g627m4vpPJymYxx34NqaMm7ki9M7d+6ko6Oj6fYGGjeQ4wtevtuyLHHt0EeASSOHDei4BJg+ffqaiOiqNq/ulb6k9wLbImKNpO4B9aAJEbEAWADQ1dUV3d0Db/Lq65dzxdrmvnQ8d8qepmMcd+Da2vCh7henV69ezUCOl4HGDeT4gpfvtixLXDv0EWDxzBEDOi7raaQn7wTeJ+kMYDjwGuAqYJSkQyJiDzAe2JzrbwYmAJskHQKMBH5XKO9TjDEzsxao+0FuRHwqIsZHxETSB7E/iYgPAbcCH8jVZgPL8/SK/Jw8/yeRxpBWAGfnu3smAZOBu4ZsTczMrK7B/ODaRUCPpHnAPcDCXL4QuC5/UNtLOlEQEQ9IWgo8COwBLoyI5wfRvpmZNamppB8Rq4HVefpR0t03lXWeBT5YI/4y0h1AZmZ2APgbuWZmJeKkb2ZWIk76ZmYl4v+cZW2t+O3UVn/7dO6UpkPMDjhf6ZuZlYiTvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYk46ZuZlYiTvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYk46ZuZlYiTvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYnUTfqShku6S9K9kh6Q9Pe5fJKkOyWtl3SDpMNy+Svz8/V5/sTCsj6Vyx+WdNp+WyszM6uqkSv954BTIuI44HhgpqRpwBeBKyPiSGA7cEGufwGwPZdfmesh6WjgbOAYYCbwdUnDhnBdzMysjrpJP5Kd+emh+RHAKcCyXL4EODNPz8rPyfNPlaRc3hMRz0XEY8B64MShWAkzM2uMIqJ+pXRFvgY4Evga8CXgjnw1j6QJwA8j4lhJ9wMzI2JTnvcIcBLwuRzzzVy+MMcsq2hrDjAHoLOzc2pPT8+AV25b7w627m4upvNwmo5x3IFv62CPa4c+tktcO/QRYNLIYXR0dDQfCEyfPn1NRHRVm3dIIwuIiOeB4yWNAm4E3jKgnjTW1gJgAUBXV1d0d3cPeFlXX7+cK9Y2tIovmjtlT9MxjjvwbR3sce3Qx3aJa4c+AiyeOYLB5L9amrp7JyKeAm4F3gGMktS3JuOBzXl6MzABIM8fCfyuWF4lxszMWqCRu3del6/wkXQ48G7gIVLy/0CuNhtYnqdX5Ofk+T+JNIa0Ajg7390zCZgM3DVE62FmZg1o5D3HWGBJHtd/BbA0Im6S9CDQI2kecA+wMNdfCFwnaT3QS7pjh4h4QNJS4EFgD3BhHjYyM7MWqZv0I+I+4G1Vyh+lyt03EfEs8MEay7oMuKz5bpqZ2VDwN3LNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSsRJ38ysRJz0zcxKxEnfzKxEnPTNzErESd/MrESc9M3MSqRu0pc0QdKtkh6U9ICkj+fyMZJWSlqX/47O5ZL0FUnrJd0n6YTCsmbn+uskzd5/q2VmZtU0cqW/B5gbEUcD04ALJR0NXAysiojJwKr8HOB0YHJ+zAGugXSSAC4FTgJOBC7tO1GYmVlr1E36EbElIn6Rp58GHgLGAbOAJbnaEuDMPD0LuDaSO4BRksYCpwErI6I3IrYDK4GZQ7kyZmbWv6bG9CVNBN4G3Al0RsSWPOtJoDNPjwM2FsI25bJa5WZm1iKKiMYqSh3AvwGXRcT3JD0VEaMK87dHxGhJNwHzI+K2XL4KuAjoBoZHxLxcfgmwOyIur2hnDmlYiM7Ozqk9PT0DXrltvTvYuru5mM7DaTrGcQe+rYM9rh362C5x7dBHgEkjh9HR0dF8IDB9+vQ1EdFVbd4hjSxA0qHAd4HrI+J7uXirpLERsSUP32zL5ZuBCYXw8blsMynxF8tXV7YVEQuABQBdXV3R3d1dWaVhV1+/nCvWNrSKL5o7ZU/TMY478G0d7HHt0Md2iWuHPgIsnjmCweS/Whq5e0fAQuChiPhyYdYKoO8OnNnA8kL5ufkunmnAjjwMdAswQ9Lo/AHujFxmZmYt0sjp553Ah4G1kn6Zyz4NzAeWSroAeBw4K8+7GTgDWA/sAs4HiIheSV8A7s71Ph8RvUOxEmZm1pi6ST+PzavG7FOr1A/gwhrLWgQsaqaDZmY2dPyNXDOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MyuRuklf0iJJ2yTdXygbI2mlpHX57+hcLklfkbRe0n2STijEzM7110mavX9Wx8zM+tPIlf5iYGZF2cXAqoiYDKzKzwFOBybnxxzgGkgnCeBS4CTgRODSvhOFmZm1Tt2kHxE/BXorimcBS/L0EuDMQvm1kdwBjJI0FjgNWBkRvRGxHVjJS08kZma2nyki6leSJgI3RcSx+flTETEqTwvYHhGjJN0EzI+I2/K8VcBFQDcwPCLm5fJLgN0RcXmVtuaQ3iXQ2dk5taenZ8Art613B1t3NxfTeThNxzjuwLd1sMe1Qx/bJa4d+ggwaeQwOjo6mg8Epk+fviYiuqrNO2RASyyIiJBU/8zR+PIWAAsAurq6oru7e8DLuvr65VyxtrlVnDtlT9MxjjvwbR3sce3Qx3aJa4c+AiyeOYLB5L9aBnr3ztY8bEP+uy2XbwYmFOqNz2W1ys3MrIUGmvRXAH134MwGlhfKz8138UwDdkTEFuAWYIak0fkD3Bm5zMzMWqjuew5J3yaNyR8haRPpLpz5wFJJFwCPA2fl6jcDZwDrgV3A+QAR0SvpC8Ddud7nI6Lyw2EzM9vP6ib9iDinxqxTq9QN4MIay1kELGqqd2ZmNqT8jVwzsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEnHSNzMrESd9M7MScdI3MysRJ30zsxJx0jczKxEnfTOzEml50pc0U9LDktZLurjV7ZuZlVlLk76kYcDXgNOBo4FzJB3dyj6YmZVZq6/0TwTWR8SjEfEHoAeY1eI+mJmVVquT/jhgY+H5plxmZmYtoIhoXWPSB4CZEfEX+fmHgZMi4qOFOnOAOfnpUcDDg2jyCOC3LYhx3IFv62CPa4c+tktcO/RxMHEAb4qI11WdExEtewDvAG4pPP8U8Kn92N7PWxHjuAPf1sEe1w59bJe4dujjYOLqPVo9vHM3MFnSJEmHAWcDK1rcBzOz0jqklY1FxB5JHwVuAYYBiyLigVb2wcyszFqa9AEi4mbg5hY1t6BFMY478G0d7HHt0Md2iWuHPg4mrl8t/SDXzMwOLP8Mg5lZibRt0pe0ON8C2kjdRZK2Sbq/GCfpOEk/k7RW0vclvaZG3PZiW5I+J2mzpF/mxxn12irMnyspJB3RYB/HSFopaV3+O7pWXEX5rZI2SnpBUled7bNO0o7KZdTblhXliyUtlPQrSQ9I+scqsZJ0m6TTK/p5TwPtHlXY3r2Sdkn6RJ2YCZKelnRefj5a0mOSJtaoe6ukB3P/P16Y9y1J99baB1WWNUzSPXn7N3qMjpK0LG+/pyR9usG4T+b+3i/pUUnnNBCzOO/H+3PsJ/qpW2t/r5D067w/fi7pxH6WsVPS43116/StVnv3FpbxY0l/UiN+uKS7cv0Hcv26+0DS0rw9XrL/++tjxWv1S3n/3SfpRkmjCjGVx9edhbgP5rKqr9V+tskNhdfEBkm/rLee0MZJvxGS+j6zWAzMrFLl/wIXR8QU4EbgbxuMA7gyIo7Pj5sbiZE0AZgBPNFEHy8GVkXEZGBVft5I3FPAl4Cf1uh/cRnrgXm16jXY3huAtwPHRcQxwOVV6gwD/gfw5fzi7ACmkvZDv21HxMN92xv4PvAcaZ/1F7OR9D2PD+Xi+cCCiNhQZd32AHMj4mhgGnCh9v5EyBRgbeU+qLKMPh8HHupvnarEXQX8KCLeQrqjbVO9OEnjgL8CuiLiWEDAOxtobxTwLtI35I8D3ivpyBr9Wkz1/T0V+E7eH58Fqp3ki9vkc3n/Vb0AaaC9+0n753jgptxmtWU8B5wSEccBx5O+/Dm5WpsVcQEsqbH/G+3jSuDYiHgr8GvSLem1jq+3AOML6/Z+Kl6r9dqLiP9SeE18F/hef+vZp22SvqRz8xn0XknX5eKTJd2er3D6zprdkv6fpBXAg5LOBb4KfJO9G/lkSbeTDty+LzA8Dfx1jtso6T7gauAzlW2RfjeoqbZy3FLgfwKvBL7fYB8/wd4vaPwa+FhfHEBE/BToBUZVbJ8dwJHA24B/qdg+D0vaATyT627N6/Tm/rZlnfaOAh4Bbs3renKNZZwAvBbYANxJOuEcVWc/bqxoayywG/h2lZjKdXsAeJukx4DzgUer9SsitkTEL/J2/nOgM++j64A3As/m/fGXwH+rtf8lPQT8PemLNfvs/36O0Y8A5+R9ex3wAtBVLw74ACmp3S3pm6QbM8ZXifsHSc/k7dILjCQd7/+aj6nfAO9vcn8DnJC3yT+TTui19gHAOxpYn/7a+yN7XxOfAybVWMaHgdsl3QssIeW44xvYJruBCXn595IumsY1mU/+N3BmLnsaOL/Qr3cDi3O/vk56ff5ZjvsB6cICYGo/26QqSQLOAr5dq84+9sfN/0P9AI4hHZxH5OdjSGe/7+SdejTpN30AuoFn8kHxYhwwkXT1VYy7B9iS475GOtvPqGjrOGB7RVu9pKT1COlgPK6Btj4KPJWX+WQTffx9xboFMKli+7ybdIVTbfusBj5YWMZ5pKRyQkXdH5CuOKpuywba6yUl2DtJ38fY2M/+eCPpCvxB4No6+7Fyf4wB1gG/qBJTa91+mrfbBQ2s2zGkE8Mm4DV5GX+o6OPz/azbCtLFxH8m/eRII8fo43l9FgP35eV8r4G4XwOfBnaSLgweqWwv19tYiBtDeof0NOmYO4GU8K7uZ5tU2983ArvysrcCG/rZB0+TXkNrgEsb2Ae1jueHcnvrgEf72SavB36Zt8vaBrfJ4kK9d5Fe16+psfxar9XivrodeLYyrpBTduZtWOzXauAjNbbJROD+GvnxZJr4Ile7XOmfQnor+VuAiOg76/1LRLwQEQ+Srsz63BURj1XGAc8X40hXV6+TtAZ4Fekt2FEVMTuqtHUY8KfAXwCbSW/na7YFDCddgQwr9PGeBvv4QsW67clxRf8R2FFt++S/jxWW8Tbgib6r2kLdH+fntbZlvfZESpTTgAtJV0mqWEbfuj4B3EB6wbxAP/uRl+6PncAE4P9Uiam1bs8AW0gv4nrrNhM4HPiriPh9jWPtFVWWcQrpImJTRKwhJbrKuFrH6L8CbwWuiTQ0sId0Z129uO+TEtQk0rufQ4HfVMSdQkomd0XEY3l9dpCuCm8B/ol0XD5fsfyiavv7KNJQ2QTgY+y96q22D35Iuug5nTSMURyPb7Q9gHm5vUXs+5tdlcfXtkhDHuNJCfrnDWwTSK+XVwFfJL3Ofl9j+VVfq33Ll/R3ed7PKuOUhjQXkY7tZXX2caPOodGrfNpoeKeG5wrTKkw/00hcRPwKeDYippLGanc22BYR8Tzp6vER0thof3F/SnphjpC0gTSkNFXSGxpobyt799OYyn7UUWv7/LFK3T/UqFtvW/bZBdweyV25rG+Io9oy+k5o/fWzWtzpwO9IQxLVYirXbQxwLOlk9En2Peb3Wb6kQ0kn8vsjojg+upt0IkDSWNJ+r7aM8cD78j7uISXij9ToZzHuadLJ4s78fAPp3VC9uInAYxHxm4j4I+kdw5sbiAP4cURMjYiT8/Nf16hXy5HAbXn6O+x7QVO5D3YBz0XENtLVbbFuo+3B3uPkevb9jlHVZUTEU6R31cWx+f62yR7S2Pj17E3ojfSxePweCryX9BlZteOrb/lPMPD8VVzmIaQT6Q2NxrRL0v8J8EFJr4V0R8tA4tj3YEPS6/PfV5CuxJ+oEjOyynKLO+gNpGGRmm1FxNqIeD2wKyImkhLWTyPiyXp9JA0X9B3gM0kHcaXbgZENbp97gLED2Jb12nuClFyR9B9yvcofi6pc18MbaKsyZjbpnUs11dZtGvDP+d3Fl0jv0l4ivytZSBqaelPFMjaS3ub3tb+nRj9fR/ogeyLw30nvLr7SwPqdAWxRukNpDOlKuN8PcnPc24F3SnpVjhtLeudZWa+blIyK+/s1+fkbScfXt/ppq9r+3kUatoB0Jdt3Aq/cB+Pz8pE0gjRc9wL9q9beqwvzZ7Hvibe4rmdL6hvv/xPStvz3KvW6eek2+UvgoYj4co1+1csnM0nH1/vYN6H/hDTEeh17h4SGyruAX0VEvePlRS3/Ru5ARMQDki4D/k3S86QDq5m4R4ARpJ30RtIBu4z0tuhVwK+AnwN/rGhrHOlD1+HANySNjIiFwGGS1ubYEcAnI2JLnbYG2sf5wCclrSON7z9SjJf0bdIBPAx4UtKTpHctbwS+AXSQxuuH55ANpPHD4rY8mfRC6pC0iX6Oi37aWwdMVLqt7A+kK7vYO8Kz734kvQvYSPocoKaK/RGkOzG+W6N65bo9kx/35flfB/5J0p9ViX0n6cS/lrTPN0naQvo8YC3w1rwPHmffd0XV+vk8KeHXVYi7hLQvniW947yygbhLSCeyvg8ie0l3kFTW+ybpbpR72fva+VtJf0O6Kn8uIp4q7qs+/ezv24FzJb0/97kvyW1g332wjnRS+E+kD7i/RRrKqqqf9qYCb5f0WdI+eMk73ryui0gfuJL7tYH0WUK9bfJ64M+AMUq3Ph6udCv2roq4/l6rX81VV5Jed4cW4pYCf5f7dG6utxlYJunPSbnkHXnb7HNRUdgmR+TX56U5D0H6/bKGh3YAfyPXzKxM2mV4x8zMhoCTvplZiTjpm5mViJO+mVmJOOmbmZWIk76ZWYk46ZuZlYiTvplZifx/iotfCsa73VgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_hs[\"chr\"].hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Yup... super-readable... Jupyter Notebooks limitations are limiting. Nonetheless, we can assume that distribution of chromosomes is uniform-ish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 237,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAWEUlEQVR4nO3df4zc9X3n8ec7OARkJ9iE3MqyfbVzdVPRWiFmBVRto3W4M4b2Yu4uRURRcTmfrJNIlOh8auCiHD1CTuQubi7orlRusTBRGoemRVhcrmTPYRvxB4Q4cXCAUC8EDq+MrWLHyQJNb7n3/TGfTQZnf8zMzsxO/Xk+pNF8v5/v5/ud9/c7s6/5zne+893ITCRJdXjTYhcgSeofQ1+SKmLoS1JFDH1JqoihL0kVWbLYBczloosuyrVr13Y07yuvvMLSpUu7W1AXWFf7BrU262qPdbVnIXUdPHjwbzPzHTNOzMyBvV166aXZqYcffrjjeXvJuto3qLVZV3usqz0LqQv4Vs6Sqy0d3omI5RHxlYj4fkQ8HRG/FhEXRsRoRBwp9ytK34iIOyNiPCKeiIiNTcvZVvofiYhtHb2FSZI61uox/c8Df5WZvwy8G3gauBk4kJnrgQNlHOBqYH257QDuAoiIC4FbgcuBy4Bbp98oJEn9MW/oR8QFwHuBuwEy8+8z84fAVmBv6bYXuLYMbwXuLZ8yHgWWR8RK4CpgNDNPZuYpYBTY0sV1kSTNI3KeyzBExCXAbuApGnv5B4GPAhOZubz0CeBUZi6PiAeBOzLzkTLtAPBxYAQ4LzNvL+2fBF7LzM+e8Xg7aHxCYGho6NJ9+/Z1tGKTk5MsW7aso3l7ybraN6i1WVd7rKs9C6lr06ZNBzNzeKZprZy9swTYCHwkMx+LiM/zs0M5AGRmRkRXLuKTmbtpvMkwPDycIyMjHS1nbGyMTuftJetq36DWZl3tsa729KquVo7pHwWOZuZjZfwrNN4EjpfDNpT7E2X6BLCmaf7VpW22dklSn8wb+pn5EvBiRLyrNF1J41DPfmD6DJxtwANleD9wQzmL5wrgdGYeAx4CNkfEivIF7ubSJknqk1Z/nPUR4IsRcS7wHHAjjTeM+yJiO/ACcF3p+1XgGmAceLX0JTNPRsSngMdLv9sy82RX1kKS1JKWQj8zDwEzfSlw5Qx9E7hpluXsAfa0UZ8kqYsG+jIMkgSw9ub/2bNl79wwxe/Nsvzn7/itnj3uYvGCa5JUEUNfkipi6EtSRQx9SaqIoS9JFfHsHUktW8hZNHOdJaP+cU9fkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5Iq4i9ye2CuXy328leJZ+O1vyV1l3v6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUkZZCPyKej4jDEXEoIr5V2i6MiNGIOFLuV5T2iIg7I2I8Ip6IiI1Ny9lW+h+JiG29WSVJ0mza2dPflJmXZOZwGb8ZOJCZ64EDZRzgamB9ue0A7oLGmwRwK3A5cBlw6/QbhSSpPxZy7Z2twEgZ3guMAR8v7fdmZgKPRsTyiFhZ+o5m5kmAiBgFtgBfWkANGhBzXW9ooea6XpHXG5La0+qefgJfi4iDEbGjtA1l5rEy/BIwVIZXAS82zXu0tM3WLknqk2jskM/TKWJVZk5ExD8CRoGPAPszc3lTn1OZuSIiHgTuyMxHSvsBGp8ARoDzMvP20v5J4LXM/OwZj7WDxmEhhoaGLt23b19HKzY5OcmyZcs6mnehDk+cnnXa0Plw/LXePO6GVRd0PO9Ct9dc67xQc22zhazzQi3ma2wuvaxrIc9zL1/7C3E2vr42bdp0sOlQ/Bu0dHgnMyfK/YmIuJ/GMfnjEbEyM4+VwzcnSvcJYE3T7KtL2wQ/Oxw03T42w2PtBnYDDA8P58jIyJldWjI2Nkan8y7UXJdO3rlhil2He3NF6+c/NNLxvAvdXr26XDTMvc0Wss4LtZivsbn0sq6FPM+9fO0vRG2vr3mfgYhYCrwpM39chjcDtwH7gW3AHeX+gTLLfuDDEbGPxpe2p8sbw0PAf2768nYzcEtX10aSuqiX31XN554tS3uy3FbedoeA+yNiuv+fZeZfRcTjwH0RsR14Abiu9P8qcA0wDrwK3AiQmScj4lPA46XfbdNf6qo7FvIC7eU/d5E0OOYN/cx8Dnj3DO0vA1fO0J7ATbMsaw+wp/0yJUnd4C9yJakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUZvAthSJrTfL+89tfVmot7+pJUEUNfkipi6EtSRQx9SaqIoS9JFTmrz95ZzH+AIEmDyD19SarIWb2nL/XS4YnTng+vf3Dc05ekihj6klQRQ1+SKmLoS1JF/CJX/6At5mm5Ozcs2kNLHXNPX5IqYuhLUkUMfUmqiKEvSRUx9CWpIi2HfkScExHfiYgHy/i6iHgsIsYj4ssRcW5pf0sZHy/T1zYt45bS/kxEXNX1tZEkzamdPf2PAk83jX8G+Fxm/iJwCthe2rcDp0r750o/IuJi4HrgV4AtwB9FxDkLK1+S1I6WQj8iVgO/BfxpGQ/gfcBXSpe9wLVleGsZp0y/svTfCuzLzJ9k5g+AceCyLqyDJKlFrf44678Bvw+8tYy/HfhhZk6V8aPAqjK8CngRIDOnIuJ06b8KeLRpmc3z/FRE7AB2AAwNDTE2NtZiiW80OTnJzg2vdzRvLw2dDzs3TM3fsc8GtS4Y3Nqsqz3W1Z7JycmO828u84Z+RPw2cCIzD0bESNcrOENm7gZ2AwwPD+fISGcPOTY2xq5HXuliZd2xc8MUuw4P3g+hB7UuGNzarKs91tWee7YspdP8m0sra/rrwPsj4hrgPOBtwOeB5RGxpOztrwYmSv8JYA1wNCKWABcALze1T2ueR5LUB/Me08/MWzJzdWaupfFF7Ncz80PAw8AHSrdtwANleH8Zp0z/emZmab++nN2zDlgPfLNrayJJmtdCPtN8HNgXEbcD3wHuLu13A1+IiHHgJI03CjLzyYi4D3gKmAJuyszBO+guSWextkI/M8eAsTL8HDOcfZOZfwf8zizzfxr4dLtFSpK6w1/kSlJFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1Jqsi8oR8R50XENyPiuxHxZET8p9K+LiIei4jxiPhyRJxb2t9SxsfL9LVNy7qltD8TEVf1bK0kSTNqZU//J8D7MvPdwCXAloi4AvgM8LnM/EXgFLC99N8OnCrtnyv9iIiLgeuBXwG2AH8UEed0cV0kSfOYN/SzYbKMvrncEngf8JXSvhe4tgxvLeOU6VdGRJT2fZn5k8z8ATAOXNaNlZAktaalY/oRcU5EHAJOAKPAs8APM3OqdDkKrCrDq4AXAcr008Dbm9tnmEeS1AdLWumUma8Dl0TEcuB+4Jd7VVBE7AB2AAwNDTE2NtbRciYnJ9m54fUuVtYdQ+fDzg1T83fss0GtCwa3Nutqj3W1Z3JysuP8m0tLoT8tM38YEQ8DvwYsj4glZW9+NTBRuk0Aa4CjEbEEuAB4ual9WvM8zY+xG9gNMDw8nCMjI22t0LSxsTF2PfJKR/P20s4NU+w63NZm74tBrQsGtzbrao91teeeLUvpNP/m0srZO+8oe/hExPnAPwOeBh4GPlC6bQMeKMP7yzhl+tczM0v79eXsnnXAeuCbXVoPSVILWnl7WwnsLWfavAm4LzMfjIingH0RcTvwHeDu0v9u4AsRMQ6cpHHGDpn5ZETcBzwFTAE3lcNGkqQ+mTf0M/MJ4D0ztD/HDGffZObfAb8zy7I+DXy6/TIlSd3gL3IlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIrMG/oRsSYiHo6IpyLiyYj4aGm/MCJGI+JIuV9R2iMi7oyI8Yh4IiI2Ni1rW+l/JCK29W61JEkzaWVPfwrYmZkXA1cAN0XExcDNwIHMXA8cKOMAVwPry20HcBc03iSAW4HLgcuAW6ffKCRJ/TFv6Gfmscz8dhn+MfA0sArYCuwt3fYC15bhrcC92fAosDwiVgJXAaOZeTIzTwGjwJZurowkaW6Rma13jlgLfAP4VeD/ZOby0h7AqcxcHhEPAndk5iNl2gHg48AIcF5m3l7aPwm8lpmfPeMxdtD4hMDQ0NCl+/bt62jFJicn+cHp1zuat5eGzofjry12FT9vUOuCwa3NutpjXe1Zd8E5LFu2rKN5N23adDAzh2eatqTVhUTEMuAvgI9l5o8aOd+QmRkRrb97zCEzdwO7AYaHh3NkZKSj5YyNjbHrkVe6UVJX7dwwxa7DLW/2vhnUumBwa7Ou9lhXe+7ZspRO828uLZ29ExFvphH4X8zMvyzNx8thG8r9idI+Aaxpmn11aZutXZLUJ62cvRPA3cDTmfmHTZP2A9Nn4GwDHmhqv6GcxXMFcDozjwEPAZsjYkX5AndzaZMk9Ukrn2l+Hfhd4HBEHCpt/wG4A7gvIrYDLwDXlWlfBa4BxoFXgRsBMvNkRHwKeLz0uy0zT3ZjJSRJrZk39MsXsjHL5Ctn6J/ATbMsaw+wp50CJUnd4y9yJakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVWTe0I+IPRFxIiK+19R2YUSMRsSRcr+itEdE3BkR4xHxRERsbJpnW+l/JCK29WZ1JElzaWVP/x5gyxltNwMHMnM9cKCMA1wNrC+3HcBd0HiTAG4FLgcuA26dfqOQJPXPvKGfmd8ATp7RvBXYW4b3Atc2td+bDY8CyyNiJXAVMJqZJzPzFDDKz7+RSJJ6bEmH8w1l5rEy/BIwVIZXAS829Tta2mZr/zkRsYPGpwSGhoYYGxvrqMDJyUl2bni9o3l7aeh82LlharHL+DmDWhcMbm3W1R7ras/k5GTH+TeXTkP/pzIzIyK7UUxZ3m5gN8Dw8HCOjIx0tJyxsTF2PfJKt8rqmp0bpth1eMGbvesGtS4Y3Nqsqz3W1Z57tiyl0/ybS6dn7xwvh20o9ydK+wSwpqnf6tI2W7skqY86Df39wPQZONuAB5rabyhn8VwBnC6HgR4CNkfEivIF7ubSJknqo3k/00TEl4AR4KKIOErjLJw7gPsiYjvwAnBd6f5V4BpgHHgVuBEgM09GxKeAx0u/2zLzzC+HJUk9Nm/oZ+YHZ5l05Qx9E7hpluXsAfa0VZ0kqav8Ra4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVJG+h35EbImIZyJiPCJu7vfjS1LN+hr6EXEO8D+Aq4GLgQ9GxMX9rEGSatbvPf3LgPHMfC4z/x7YB2ztcw2SVK3IzP49WMQHgC2Z+W/K+O8Cl2fmh5v67AB2lNF3Ac90+HAXAX+7gHJ7xbraN6i1WVd7rKs9C6nrFzLzHTNNWNJ5Pb2RmbuB3QtdTkR8KzOHu1BSV1lX+wa1Nutqj3W1p1d19fvwzgSwpml8dWmTJPVBv0P/cWB9RKyLiHOB64H9fa5BkqrV18M7mTkVER8GHgLOAfZk5pM9ergFHyLqEetq36DWZl3tsa729KSuvn6RK0laXP4iV5IqYuhLUkXOytBfzEs9RMSaiHg4Ip6KiCcj4qOl/Q8iYiIiDpXbNU3z3FJqfSYiruphbc9HxOHy+N8qbRdGxGhEHCn3K0p7RMSdpa4nImJjj2p6V9M2ORQRP4qIjy3G9oqIPRFxIiK+19TW9vaJiG2l/5GI2Najuv5rRHy/PPb9EbG8tK+NiNeattsfN81zaXn+x0vt0YO62n7eevH3OkttX26q6/mIOFTa+7LN5siG/r7GMvOsutH4gvhZ4J3AucB3gYv7+PgrgY1l+K3A39C45MQfAP9+hv4XlxrfAqwrtZ/To9qeBy46o+2/ADeX4ZuBz5Tha4D/BQRwBfBYn567l4BfWIztBbwX2Ah8r9PtA1wIPFfuV5ThFT2oazOwpAx/pqmutc39zljON0utUWq/ugd1tfW89ervdabazpi+C/iP/dxmc2RDX19jZ+Oe/qJe6iEzj2Xmt8vwj4GngVVzzLIV2JeZP8nMHwDjNNahX7YCe8vwXuDapvZ7s+FRYHlErOxxLVcCz2bmC3P06dn2ysxvACdneLx2ts9VwGhmnszMU8AosKXbdWXm1zJzqow+SuM3L7Mqtb0tMx/NRnLc27QuXatrDrM9bz35e52rtrK3fh3wpbmW0e1tNkc29PU1djaG/irgxabxo8wduj0TEWuB9wCPlaYPl49pe6Y/wtHfehP4WkQcjMblLgCGMvNYGX4JGFqEuqZdzxv/EBd7e0H722cxttu/prFHOG1dRHwnIv46In6ztK0qtfSjrnaet8XYXr8JHM/MI01tfd1mZ2RDX19jZ2PoD4SIWAb8BfCxzPwRcBfwT4BLgGM0Pl72229k5kYaVzm9KSLe2zyx7M0syjm80fix3vuBPy9Ng7C93mAxt89sIuITwBTwxdJ0DPjHmfke4N8BfxYRb+tjSQP3vM3gg7xx56Kv22yGbPipfrzGzsbQX/RLPUTEm2k8qV/MzL8EyMzjmfl6Zv4/4E/42SGJvtWbmRPl/gRwf6nh+PRhm3J/ot91FVcD387M46XGRd9eRbvbp2/1RcTvAb8NfKiEBeXwyctl+CCN4+W/VGpoPgTUk7o6eN76+nxGxBLgXwJfbqq5b9tspmygz6+xszH0F/VSD+V44d3A05n5h03tzcfD/wUwfVbBfuD6iHhLRKwD1tP48qjbdS2NiLdOD9P4IvB75fGnv/3fBjzQVNcN5QyCK4DTTR9Be+ENe1+Lvb2atLt9HgI2R8SKcmhjc2nrqojYAvw+8P7MfLWp/R3R+L8VRMQ7aWyf50ptP4qIK8pr9IamdelmXe0+b/3+e/2nwPcz86eHbfq1zWbLBvr9Guv0m+hBvtH41vtvaLxjf6LPj/0bND6ePQEcKrdrgC8Ah0v7fmBl0zyfKLU+wwLPqJijrnfSODPiu8CT09sFeDtwADgC/G/gwtIeNP7hzbOl7uEebrOlwMvABU1tfd9eNN50jgH/l8Zx0u2dbB8ax9jHy+3GHtU1TuO47vRr7I9L339Vnt9DwLeBf960nGEaIfws8N8pv8jvcl1tP2+9+HudqbbSfg/wb8/o25dtxuzZ0NfXmJdhkKSKnI2HdyRJszD0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkX+Pw2I1ANoVICBAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_hs[\"pos\"].hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Problem... `pos` column contains integer objects, so Pandas aggregates them in integer bins instead of treating them like unique values. Let's convert them to strings before making a histogram and see how it will works!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 238,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAARFUlEQVR4nO3dUYxc1X3H8e+fNSRV09omJCtkuzVSrEZEEQnaGKJU1TZWjaGo5iFBRFXZIqv74kapVKklfbEKQUpeSoOUoG6LGxOlIRZtvFaEQi0nq6gPEKBJSYBQbyHItgC3WZsmQpB6+fdhzoZhM+OZnZ29u/b5fqTVPfec8797Lki/ub5zZzYyE0lSHS5a7QVIkppj6EtSRQx9SaqIoS9JFTH0Jaki61Z7Aedy2WWX5datWweuP3PmzNDWIklNigjWr18/UO0TTzzxP5n5rk5jazr0t27dyuOPPz5w/fT09BBXI0nNGRkZ4cYbbxyoNiJe6Dbm7R1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SarImv5E7vnqjYQAFv48zUK7U1+v8aXUXBTDPhNJFxqv9CWpIoa+JFXE0JekilzQ9/QzB7+X3mv8XDWStFZd0KFfm+W8yC23xjeRpfODt3ckqSKGviRVxNCXpIoY+pJUkb5CPyI2RMSDEfGjiHgmIj4cEZdGxJGIOFa2G8vciIh7ImI2Ip6MiKvbjjNR5h+LiImVOilJUmf9Xul/HvhmZr4XuAp4BrgdOJqZ24CjZR/gemBb+ZkE7gWIiEuBfcA1wHZg38ILhSSpGT1DPyLWA78D3AeQmT/PzDPAbuBAmXYAuKm0dwP3Z8sjwIaIuBy4DjiSmXOZeRo4Auwa4rlIknro5zn9K4D/Bv4xIq4CngA+BYxm5otlzkvAaGlvAo631Z8ofd363yIiJmn9C4HR0VFmZmb6PZeOYtG2W3uQcWv8MJq0Uubn55edf530E/rrgKuBT2bmoxHxed68lQNAZmZEZMfqJcrMKWAKYGxsLMfHxwc+1qFD06vyidxaayQNz8jICMvJv276uad/AjiRmY+W/QdpvQi8XG7bULanyvhJYEtb/ebS161fktSQnqGfmS8BxyPit0rXDuBp4DCw8ATOBDBd2oeBW8tTPNcCr5TbQA8DOyNiY3kDd2fpkyQ1pN/v3vkk8JWIuAR4DriN1gvGwYjYA7wA3FzmPgTcAMwCr5a5ZOZcRNwJPFbm3ZGZc0M5C0lSX/oK/cz8PjDWYWhHh7kJ7O1ynP3A/iWsT5I0RH4iV5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIr0+0dUJC3yRjb394jb2xf5B4m1DF7pS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUkb5CPyJ+HBE/iIjvR8Tjpe/SiDgSEcfKdmPpj4i4JyJmI+LJiLi67TgTZf6xiJhYmVOSJHWzlCv9383MD2TmWNm/HTiamduAo2Uf4HpgW/mZBO6F1osEsA+4BtgO7Ft4oZAkNWM5t3d2AwdK+wBwU1v//dnyCLAhIi4HrgOOZOZcZp4GjgC7lvH7JUlL1O/XMCTwrxGRwN9l5hQwmpkvlvGXgNHS3gQcb6s9Ufq69b9FREzS+hcCo6OjzMzM9LnEzmLRtlt7kHFr3tq3GjJ7z1kpTf+3XtzWhW1+fn7Z+ddJv6H/25l5MiLeDRyJiB+1D2ZmlheEZSsvKFMAY2NjOT4+PvCxDh2aHsp3nVjTX99qWGv/DZr47h2Dvw4jIyMsJ/+66ev2TmaeLNtTwNdp3ZN/udy2oWxPlekngS1t5ZtLX7d+SVJDeoZ+RPxqRPzaQhvYCfwQOAwsPIEzAUyX9mHg1vIUz7XAK+U20MPAzojYWN7A3Vn6JEkN6ef2zijw9YhYmP9PmfnNiHgMOBgRe4AXgJvL/IeAG4BZ4FXgNoDMnIuIO4HHyrw7MnNuaGciSeqpZ+hn5nPAVR36fwLs6NCfwN4ux9oP7F/6MiVJw+AnciWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JF+v0aBumcMlfnawskLY1X+pJUEUNfkiri7R1Jfet0G6+9vVK39MJ7eUPjlb4kVcTQl6SKGPqSVBHv6UvnmX4fj+01PmiNzm9e6UtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkX6Dv2IGImI70XEN8r+FRHxaETMRsTXIuKS0v+2sj9bxre2HePTpf/ZiLhu6GcjSTqnpVzpfwp4pm3/c8Ddmfke4DSwp/TvAU6X/rvLPCLiSuAW4H3ALuCLETGyvOVLkpair9CPiM3A7wP/UPYD+CjwYJlyALiptHeXfcr4jjJ/N/BAZr6emc8Ds8D2IZyDJKlP/V7p/y3wF8AbZf+dwJnMPFv2TwCbSnsTcBygjL9S5v+iv0ONJKkBPb9lMyJuBE5l5hMRMb7SC4qISWASYHR0lJmZmeUdb9G2W3uQcWusabJmraxjNWpqND8/v+z866Sfr1b+CPAHEXED8Hbg14HPAxsiYl25mt8MnCzzTwJbgBMRsQ5YD/ykrX9Be80vZOYUMAUwNjaW4+PjA5xWy6FD0wN/hWyvcWusabJmraxjtWpqNDIywnLyr5uet3cy89OZuTkzt9J6I/ZbmfmHwLeBj5VpE8B0aR8u+5Txb2Vmlv5bytM9VwDbgO8O7UwkXbDeyNZPLtr26ltuzYVoOX9E5S+BByLiM8D3gPtK/33AlyNiFpij9UJBZj4VEQeBp4GzwN7MnF/G75ckLdGSQj8zZ4CZ0n6ODk/fZOZrwMe71N8F3LXURUqShsNP5EpSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkirSM/Qj4u0R8d2I+I+IeCoi/rr0XxERj0bEbER8LSIuKf1vK/uzZXxr27E+XfqfjYjrVuysJEkd9XOl/zrw0cy8CvgAsCsirgU+B9ydme8BTgN7yvw9wOnSf3eZR0RcCdwCvA/YBXwxIkaGeC6SpB56hn62/KzsXlx+Evgo8GDpPwDcVNq7yz5lfEdEROl/IDNfz8zngVlg+zBOQpLUn3X9TCpX5E8A7wG+APwXcCYzz5YpJ4BNpb0JOA6QmWcj4hXgnaX/kbbDtte0/65JYBJgdHSUmZmZpZ3R4uMt2nZrDzJujTVN1qyVddRWs1rm5+eXnX+d9BX6mTkPfCAiNgBfB9479JW8+bumgCmAsbGxHB8fH/hYhw5Nk7T+52Xp69QeZNwaa5qsWSvrqLFmtYyMjLCc/OtmSU/vZOYZ4NvAh4ENEbHworEZOFnaJ4EtAGV8PfCT9v4ONZKkBvTz9M67yhU+EfErwO8Bz9AK/4+VaRPAdGkfLvuU8W9lZpb+W8rTPVcA24DvDuk8JEl96Of2zuXAgXJf/yLgYGZ+IyKeBh6IiM8A3wPuK/PvA74cEbPAHK0ndsjMpyLiIPA0cBbYW24bSZIa0jP0M/NJ4IMd+p+jw9M3mfka8PEux7oLuGvpy5QkDYOfyJWkihj6klQRQ1+SKmLoS1JFDH1JqoihL0kV6etrGCSpRpmr9xURK/UVxF7pS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5Iq0jP0I2JLRHw7Ip6OiKci4lOl/9KIOBIRx8p2Y+mPiLgnImYj4smIuLrtWBNl/rGImFi505IkddLPlf5Z4M8z80rgWmBvRFwJ3A4czcxtwNGyD3A9sK38TAL3QutFAtgHXANsB/YtvFBIkprRM/Qz88XM/PfS/inwDLAJ2A0cKNMOADeV9m7g/mx5BNgQEZcD1wFHMnMuM08DR4BdwzwZSdK5Lelv5EbEVuCDwKPAaGa+WIZeAkZLexNwvK3sROnr1r/4d0zS+hcCo6OjzMzMLGWJv7zmRdtu7UHGrbGmyZq1sg5rmqmZn59fdv510nfoR8Q7gH8G/iwz/zfizSVmZkZEdi1egsycAqYAxsbGcnx8fOBjHTo0PfAfJe41bo01TdaslXVY01zNupERlpN/3fT19E5EXEwr8L+Smf9Sul8ut20o21Ol/ySwpa18c+nr1i9Jakg/T+8EcB/wTGb+TdvQYWDhCZwJYLqt/9byFM+1wCvlNtDDwM6I2FjewN1Z+iRJDenn9s5HgD8CfhAR3y99fwV8FjgYEXuAF4Cby9hDwA3ALPAqcBtAZs5FxJ3AY2XeHZk5N4yTkCT1p2foZ+a/8db3GNrt6DA/gb1djrUf2L+UBUqShsdP5EpSRQx9SaqIoS9JFTH0Jakihr4kVcTQl6SKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUkZ6hHxH7I+JURPywre/SiDgSEcfKdmPpj4i4JyJmI+LJiLi6rWaizD8WERMrczqSpHPp50r/S8CuRX23A0czcxtwtOwDXA9sKz+TwL3QepEA9gHXANuBfQsvFJKk5vQM/cz8DjC3qHs3cKC0DwA3tfXfny2PABsi4nLgOuBIZs5l5mngCL/8QiJJWmHrBqwbzcwXS/slYLS0NwHH2+adKH3d+n9JREzS+lcCo6OjzMzMDLjEcrxF227tQcatsabJmrWyDmuaqZmfn192/nUyaOj/QmZmROQwFlOONwVMAYyNjeX4+PjAxzp0aJqk9R9wYYGd2oOMW2NNkzVrZR3WNFezbmSE5eRfN4M+vfNyuW1D2Z4q/SeBLW3zNpe+bv2SpAYNGvqHgYUncCaA6bb+W8tTPNcCr5TbQA8DOyNiY3kDd2fpkyQ1qOftnYj4KjAOXBYRJ2g9hfNZ4GBE7AFeAG4u0x8CbgBmgVeB2wAycy4i7gQeK/PuyMzFbw5LklZYz9DPzE90GdrRYW4Ce7scZz+wf0mrkyQNlZ/IlaSKGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIoa+JFXE0Jekihj6klQRQ1+SKmLoS1JFDH1JqoihL0kVMfQlqSKGviRVxNCXpIoY+pJUEUNfkipi6EtSRQx9SaqIoS9JFTH0Jakihr4kVaTx0I+IXRHxbETMRsTtTf9+SapZo6EfESPAF4DrgSuBT0TElU2uQZJq1vSV/nZgNjOfy8yfAw8AuxtegyRVa13Dv28TcLxt/wRwTfuEiJgEJsvuzyLi2WX8vvcD/wdcXLZ0aQ8ybo01TdaslXVY01zN68DzDOY3uw00Hfo9ZeYUMDWMY0VE0jrHi3jzXDu1Bxm3xpoma9bKOqxpriYyc4wha/r2zklgS9v+5tInSWpA06H/GLAtIq6IiEuAW4DDDa9BkqrV6O2dzDwbEX8KPAyMAPsz86kV/JXPA6eAd5ctXdqDjFtjTZM1a2Ud1jRXc4wVEJm5EseVJK1BfiJXkipi6EtSRdbcI5vnEhH7gdtWex2StAa8CrwA3JCZP+636LwKfeBLtB75/BCwHpgDNgJv0HpjWJIudAn8FPgO8CBvvgHcl/Pq9k5mfgf4E+BnpWthG6uzIklqXNC6wt8IvJaZry6l+LwK/XO4UM5DkvrxfuCDwIfKF1n2zbCUpPPPfPm5BvjjpRReCKHvrR1JtQla72O+Bly9lMILIfSTN7+dTpIuRIs/RRu0voXzHcDTSznQeRX6EfFVWl+tsKl0/Qatk7941RYlSStv8R2NKD//Cfz9kg7k1zBIUj3Oqyt9SdLyGPqSVBFDX5IqYuhLUkUMfUmqiKEvSRUx9CWpIv8Pb8SJ7Yi/4Z0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_hs[\"pos\"].astype(str).hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Oh, well... maybe an integer histogram wasn't that bad after all? Again, distribution is uniform-ish"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 239,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAY6klEQVR4nO3df5BdZX3H8ffHRAhmIQk/utIkNbFGLZJizQ5Q0XZXbNigbXCKCEMlMNiMNnXaaZyCtQz+AI0d0FFUakZigqJrBuskjQiTCVmVTvkViYSANMsPJRETZWNsJGqj3/5xnpXrcnfvub82N3k+r5k7e89znvM9z/ece7/33HPPvauIwMzM8vCCQz0AMzObOC76ZmYZcdE3M8uIi76ZWUZc9M3MMjL5UA9gPCeeeGLMmTOn4eV//vOfM3Xq1JaNJ7d47YjZ6fHaETO3eO2I2enx2hGzmXhbtmz5SUScVHVmRHTsbcGCBdGMzZs3N7V87vHaEbPT47UjZm7x2hGz0+O1I2Yz8YD7Y4y66tM7ZmYZcdE3M8uIi76ZWUZc9M3MMuKib2aWERd9M7OMuOibmWXERd/MLCMu+mZmGenon2Gw+mzbtY9Lr/x6S2Mun3+wVMwnV7yppes1s/bwkb6ZWUZc9M3MMuKib2aWERd9M7OM+INcs8NMqz+wL/thPfgD+yOBj/TNzDLiI307rPkyVbP6+EjfzCwjpYq+pOmSbpX0PUmPSPpTScdL2ihpR/o7I/WVpE9KGpL0oKTXVMRZkvrvkLSkXUmZmVl1ZY/0PwHcHhGvBE4DHgGuBDZFxDxgU5oGWATMS7elwI0Ako4HrgbOAE4Hrh55oTAzs4lR85y+pGnAnwGXAkTEr4BfSVoM9KZua4BB4ApgMXBz+ue8d6d3CSenvhsjYjjF3Qj0A19uXTpmdiQ6VFcsHYmf25Q50p8L/Bj4vKQHJH1O0lSgOyKeTn1+BHSn+zOBpyqW35naxmo3M7MJouKAfJwOUg9wN3BWRNwj6RPAz4B3R8T0in57I2KGpA3Aioi4K7VvongH0AtMiYhrUvtVwIGIuG7U+pZSnBaiu7t7wcDAQMPJ7d+/n66uroaXP9zi7Rnex+4DLQsHQPcxlIo5f+a0UvGcc/NanXPZfCG/nMvmC51VH/r6+rZERE+1eWUu2dwJ7IyIe9L0rRTn73dLOjkink6nb/ak+buA2RXLz0ptu3judNBI++DolUXESmAlQE9PT/T29o7uUtrg4CDNLH+4xbvhlnVcv621V+Eun3+wVMwnL+4tFc85N6/VOZfNF/LLuWy+0Pn1YUTN0zsR8SPgKUmvSE1nAw8D64GRK3CWAOvS/fXAJekqnjOBfek00B3AQkkz0ge4C1ObmZlNkLIvne8GbpF0FPA4cBnFC8ZaSZcD3wcuSH1vA84FhoBnU18iYljSh4D7Ur8Pjnyoa2ZmE6NU0Y+IrUC180NnV+kbwLIx4qwCVtUxPjMzayF/I9fMLCMu+mZmGfEPrpmZjWFOHV8Iq+cnqstY3T+1ZbEqHdFF39/iMzP7XT69Y2aWERd9M7OMuOibmWXERd/MLCMu+mZmGXHRNzPLiIu+mVlGXPTNzDLiom9mlhEXfTOzjLjom5llxEXfzCwjLvpmZhlx0Tczy4iLvplZRlz0zcwy4qJvZpYRF30zs4y46JuZZcRF38wsI6WKvqQnJW2TtFXS/anteEkbJe1If2ekdkn6pKQhSQ9Kek1FnCWp/w5JS9qTkpmZjaWeI/2+iHh1RPSk6SuBTRExD9iUpgEWAfPSbSlwIxQvEsDVwBnA6cDVIy8UZmY2MZo5vbMYWJPurwHOq2i/OQp3A9MlnQycA2yMiOGI2AtsBPqbWL+ZmdVJEVG7k/QEsBcI4LMRsVLSTyNiepovYG9ETJe0AVgREXeleZuAK4BeYEpEXJParwIORMR1o9a1lOIdAt3d3QsGBgYaTm7P8D52H2h48efpPoZS8ebPnFYq3v79++nq6mpyVM9pdb7gnMdzpORcNl/IM+dDFXPutEkNb8O+vr4tFWdlfsfkkjFeFxG7JP0esFHS9ypnRkRIqv3qUUJErARWAvT09ERvb2/DsW64ZR3XbyubYm3L5x8sFe/Ji3tLxRscHKSZ/EZrdb7gnMdzpORcNl/IM+dDFXN1/9SWbsMRpU7vRMSu9HcP8DWKc/K702kb0t89qfsuYHbF4rNS21jtZmY2QWoWfUlTJR07ch9YCDwErAdGrsBZAqxL99cDl6SreM4E9kXE08AdwEJJM9IHuAtTm5mZTZAy70W6ga8Vp+2ZDHwpIm6XdB+wVtLlwPeBC1L/24BzgSHgWeAygIgYlvQh4L7U74MRMdyyTMzMrKaaRT8iHgdOq9L+DHB2lfYAlo0RaxWwqv5hmplZK/gbuWZmGXHRNzPLiIu+mVlGXPTNzDLiom9mlhEXfTOzjLjom5llxEXfzCwjLvpmZhlx0Tczy4iLvplZRlz0zcwy4qJvZpYRF30zs4y46JuZZcRF38wsIy76ZmYZcdE3M8uIi76ZWUZc9M3MMuKib2aWERd9M7OMuOibmWWkdNGXNEnSA5I2pOm5ku6RNCTpK5KOSu1Hp+mhNH9ORYz3pvZHJZ3T8mzMzGxc9Rzp/wPwSMX0R4GPR8TLgL3A5an9cmBvav946oekU4ALgVcB/cBnJE1qbvhmZlaPUkVf0izgTcDn0rSANwC3pi5rgPPS/cVpmjT/7NR/MTAQEb+MiCeAIeD0FuRgZmYlKSJqd5JuBT4CHAu8B7gUuDsdzSNpNvCNiDhV0kNAf0TsTPMeA84A3p+W+WJqvyktc+uodS0FlgJ0d3cvGBgYaDi5PcP72H2g4cWfp/sYSsWbP3NaqXj79++nq6uryVE9p9X5gnMez5GSc9l8Ic+cD1XMudMmNbwN+/r6tkRET7V5k2stLOnNwJ6I2CKpt6ER1CEiVgIrAXp6eqK3t/FV3nDLOq7fVjPF0pbPP1gq3pMX95aKNzg4SDP5jdbqfME5j+dIyblsvpBnzocq5ur+qS3dhiPKjPAs4K8knQtMAY4DPgFMlzQ5Ig4Cs4Bdqf8uYDawU9JkYBrwTEX7iMplzMxsAtQ8px8R742IWRExh+KD2Dsj4mJgM3B+6rYEWJfur0/TpPl3RnEOaT1wYbq6Zy4wD7i3ZZmYmVlNzbwXuQIYkHQN8ABwU2q/CfiCpCFgmOKFgojYLmkt8DBwEFgWEb9uYv1mZlanuop+RAwCg+n+41S5+iYifgG8dYzlrwWurXeQZmbWGv5GrplZRlz0zcwy4qJvZpYRF30zs4y46JuZZcRF38wsIy76ZmYZae2PT5jZEW3OlV8v1W/5/INcWrJvuXgtC5U9H+mbmWXERd/MLCMu+mZmGXHRNzPLiIu+mVlGfPVOG/gKBzPrVD7SNzPLiIu+mVlGXPTNzDLiom9mlhF/kGvWIH9gb4cjH+mbmWXERd/MLCMu+mZmGXHRNzPLSM2iL2mKpHslfVfSdkkfSO1zJd0jaUjSVyQdldqPTtNDaf6ciljvTe2PSjqnbVmZmVlVZY70fwm8ISJOA14N9Es6E/go8PGIeBmwF7g89b8c2JvaP576IekU4ELgVUA/8BlJk1qYi5mZ1VCz6Edhf5p8YboF8Abg1tS+Bjgv3V+cpknzz5ak1D4QEb+MiCeAIeD0ViRhZmbllDqnL2mSpK3AHmAj8Bjw04g4mLrsBGam+zOBpwDS/H3ACZXtVZYxM7MJoIgo31maDnwNuApYnU7hIGk28I2IOFXSQ0B/ROxM8x4DzgDeD9wdEV9M7TelZW4dtY6lwFKA7u7uBQMDAw0nt2d4H7sPNLz483QfQ1bx6ok5f+a0UvH2799PV1dXk6N6Tqv3MXT+fun0eO2I2enx2hFz7rRJDT9X+vr6tkRET7V5dX0jNyJ+Kmkz8KfAdEmT09H8LGBX6rYLmA3slDQZmAY8U9E+onKZynWsBFYC9PT0RG9vbz1D/B033LKO67e17kvHy+cfzCpePTGfvLi3VLzBwUGa2aejtXofQ+fvl06P146YnR6vHTFX909t6XNlRJmrd05KR/hIOgb4C+ARYDNwfuq2BFiX7q9P06T5d0bxdmI9cGG6umcuMA+4t0V5mJlZCWVelk4G1qQrbV4ArI2IDZIeBgYkXQM8ANyU+t8EfEHSEDBMccUOEbFd0lrgYeAgsCwift3adMzMbDw1i35EPAj8SZX2x6ly9U1E/AJ46xixrgWurX+YZmbWCv5GrplZRvzTytYS/plhs8ODj/TNzDLiom9mlhEXfTOzjLjom5llxEXfzCwjLvpmZhlx0Tczy4iLvplZRlz0zcwy4qJvZpYRF30zs4y46JuZZcRF38wsIy76ZmYZcdE3M8uIi76ZWUZc9M3MMuKib2aWERd9M7OMuOibmWXERd/MLCMu+mZmGalZ9CXNlrRZ0sOStkv6h9R+vKSNknakvzNSuyR9UtKQpAclvaYi1pLUf4ekJe1Ly8zMqilzpH8QWB4RpwBnAssknQJcCWyKiHnApjQNsAiYl25LgRuheJEArgbOAE4Hrh55oTAzs4lRs+hHxNMR8Z10/3+BR4CZwGJgTeq2Bjgv3V8M3ByFu4Hpkk4GzgE2RsRwROwFNgL9rUzGzMzGp4go31maA3wLOBX4QURMT+0C9kbEdEkbgBURcVeatwm4AugFpkTENan9KuBARFw3ah1LKd4h0N3dvWBgYKDh5PYM72P3gYYXf57uY8gqXjtidnq8dsTMLV47YnZ6vHbEnDttEl1dXQ0t29fXtyUieqrNm1w2iKQu4KvAP0bEz4o6X4iIkFT+1WMcEbESWAnQ09MTvb29Dce64ZZ1XL+tdIo1LZ9/MKt47YjZ6fHaETO3eO2I2enx2hFzdf9Umql/Yyl19Y6kF1IU/Fsi4j9S8+502ob0d09q3wXMrlh8Vmobq93MzCZImat3BNwEPBIRH6uYtR4YuQJnCbCuov2SdBXPmcC+iHgauANYKGlG+gB3YWozM7MJUua9yFnA24Ftkramtn8BVgBrJV0OfB+4IM27DTgXGAKeBS4DiIhhSR8C7kv9PhgRw61IwszMyqlZ9NMHshpj9tlV+gewbIxYq4BV9QzQzMxax9/INTPLiIu+mVlGXPTNzDLiom9mlhEXfTOzjLjom5llxEXfzCwjLvpmZhlx0Tczy4iLvplZRlz0zcwy4qJvZpYRF30zs4y46JuZZcRF38wsIy76ZmYZcdE3M8uIi76ZWUZc9M3MMuKib2aWERd9M7OMuOibmWXERd/MLCM1i76kVZL2SHqoou14SRsl7Uh/Z6R2SfqkpCFJD0p6TcUyS1L/HZKWtCcdMzMbT5kj/dVA/6i2K4FNETEP2JSmARYB89JtKXAjFC8SwNXAGcDpwNUjLxRmZjZxahb9iPgWMDyqeTGwJt1fA5xX0X5zFO4Gpks6GTgH2BgRwxGxF9jI819IzMyszRQRtTtJc4ANEXFqmv5pRExP9wXsjYjpkjYAKyLirjRvE3AF0AtMiYhrUvtVwIGIuK7KupZSvEugu7t7wcDAQMPJ7Rnex+4DDS/+PN3HkFW8dsTs9HjtiJlbvHbE7PR47Yg5d9okurq6Glq2r69vS0T0VJs3ualRARERkmq/cpSPtxJYCdDT0xO9vb0Nx7rhlnVcv63pFH9r+fyDWcVrR8xOj9eOmLnFa0fMTo/Xjpir+6fSTP0bS6NX7+xOp21If/ek9l3A7Ip+s1LbWO1mZjaBGi3664GRK3CWAOsq2i9JV/GcCeyLiKeBO4CFkmakD3AXpjYzM5tANd+LSPoyxTn5EyXtpLgKZwWwVtLlwPeBC1L324BzgSHgWeAygIgYlvQh4L7U74MRMfrDYTMza7OaRT8iLhpj1tlV+gawbIw4q4BVdY3OzMxayt/INTPLiIu+mVlGXPTNzDLiom9mlhEXfTOzjLjom5llxEXfzCwjLvpmZhlx0Tczy4iLvplZRlz0zcwy4qJvZpYRF30zs4y46JuZZcRF38wsIy76ZmYZcdE3M8uIi76ZWUZc9M3MMuKib2aWERd9M7OMuOibmWXERd/MLCMu+mZmGZnwoi+pX9KjkoYkXTnR6zczy9mEFn1Jk4BPA4uAU4CLJJ0ykWMwM8vZRB/pnw4MRcTjEfErYABYPMFjMDPLliJi4lYmnQ/0R8Q70vTbgTMi4u8r+iwFlqbJVwCPNrHKE4GfNLF87vHaEbPT47UjZm7x2hGz0+O1I2Yz8V4SESdVmzG58fG0R0SsBFa2Ipak+yOipxWxcozXjpidHq8dMXOL146YnR6vHTHbMUaY+NM7u4DZFdOzUpuZmU2AiS769wHzJM2VdBRwIbB+gsdgZpatCT29ExEHJf09cAcwCVgVEdvbuMqWnCbKOF47YnZ6vHbEzC1eO2J2erx2xGzHGCf2g1wzMzu0/I1cM7OMuOibmeUkIg7rG3AeEMArR7V/A5iV7p8I/B/wznHinABsTbcfUVxVNDJ9FPDvwFmp7z8B3wO2Ad8FPga8cJzYL6b4ItpjwBbgNuDlFePcnNYzBOyrWO9rmxjje9IYt1J8gH5JE+OrZzt2A18CHk+x/ht4S8X8LcDR6f6r077rr7GPy8R8IOX6A+DHFdtlThPbcHKKtaLG+GpuvxTrw8COivW9r9ntCHQBn61Y9yDFd1/qfq4ALwRWpDF+J61zUaPPvXpyrjPfG1O+30ntf9vEfimdc8nHzshz+WHgQMX88xvcjqVqQ101s9EFO+UGfAX4NvCBirZjgHsrpt+V+nyzZMz3A+8Z1baV4sPndwK3A9NT+1HAlcBxY8RSeiC9s6LtNOD1VcbZC2xo0RjvGBkTcBywpAXjG3c7jhHrJcC70/25wPqKeR9N8daMk2e9MS8FPtXsNkz3FwH/RVE01Mz2oygsq4EpafpY4P3N5kxR1D4CvKBi3psaea6kMa7huRflbuCCRp97ZXNuIN8PV+R7EnBFk/uldM51PHbmAA+VfByWqWG9lKwNNdfXiiCH6kbxqr8LeDnwaEX7IuDfKqa/RfoJCNJRaz07FPgjYG26/xQwt44xvgH41hjzRo+z9I6tMcYfAC9tw/jG3Y7A2YzzwkrxovF36b4ojur+EPjhSGFoJmaavpQGi37lNkzTNwMXUBxtVT2yKrP9gBcBzwDHlhxXqZzTtntipNDUiDnuc6VijFUPXpqIVzPnOvN9nFTwm31c15tznY+dOZQo+rW2Y8V0Ly0q+of7Of3FwO0R8T/AM5IWpPZFFEfjSJoNnBwR9wJrgbc1sJ5FwO2SjgO6IuKJOpY9leKt5ZhxGxjPmLHSGI+NiMdLLldqfCW346so3iKPpZ/n8n0t8EREPEZxSuJNYyxTT8xmVeY7BXgj8J/Al4GLxlimzPZ7GfCDiPjfkuMom/OrgK0R8esSMWs9V0bG+LOSYywbr0zO9eT73Yj4TYmY9eyXsjmPp9Hncs0a1mqHe9G/iOLtHunvyBPzLOCudP9tFEVqdJ96nEOVHSDpHElbJT0p6bUNxK0cZ7OqjrFJTW1HSZ+W9F1J96Uv482qeDEaa981E7NZldvwzcDmiDgAfBU4L/1KbD2q7l9Jl6XHzVPpxXRcLcq5zHOlbfHqyblsvpLel2L+sM6xt/J5N6LR51+r90tNHffbO2VJOp7iLdx8SUFxLjsk3Qg8FcWveEKxEV8s6eI0/fuS5kXEjpLreRHF+fsfpun9kuZGxBMRcQdwh6QNFOf2q9kOnF8l7ktHjbNhY4zxpSULQ9nxldmO24G/HpmIiGWSTgTupziXeleKPSn1WyzpfRSnek6QdGyVI8NSMZs1ehumfF8n6ck0fQLF421jlfGNu/0kDQF/MJJfRHwe+Lykhyget6OVzXk7cJqkSeMd7Zd5rlSM8bhaR751xiuTc9l8H075viAifhMR1wLXSto/xjYsu19q5lxje4x+7JRdrmwNa6nD+Uj/fOALEfGSiJgTEbMpzm/+Dc+9RX85xemYmanPHIoPveo52u+jOKc74iPAjZKmp3UImDLO8ncCR6dfDyUt88fA22ndkXm1MX46nepBUpekSxodXx3b8U5giqR3VbS9KP3tp7gaAYpzuA9GxOwU7yUUR9NvGWN8ZWI267fbMG231wN/UJHvMqo/bmpuv4h4FrgJ+FQ6bTTywjfWgUKpnNOpsfuBD6THIZLmSBp9qqzmc6VijJ9IR9dIOknSW6uMr554ZXIum+9QyveakXddKbbGiFl2v5TJeTyjn39l1dyO7XA4F/2LgK+NavsqcCbPbbCx+tRT9EefW7sR2ATcI+lBiqs7Hki354niU5i3AG+U9Jik7RQFs3Kczao2xs3AfenI6ttA1fOgJcdXajumWOcBfy7pCUn3UlwZcQXFB1HfrCdenTGbVbkN3wLcGRG/rJi/DvhLSUdXGV+Z/fs+4GngIUkPkK5aovgQ+3fUmfM7KK44GUr7ejWwZ1TIMs8VgH+luET14RRrA1DtCLhsvFI5N5DvCSnf+yneef3z6AHWsV/K5jyeRs+/l92OrdWKT4M75UZxHe/9LY75Hca5Br8TxtnqMbZhfLOAb7R4G7Y0Ziu3YTseh63OuQ37uB3PvVbv43btl45/7FTe/Ns7ZmYZOZxP75iZWZ1c9M3MMuKib2aWERd9M7OMuOibmWXERd/MLCP/D3Mt+xyDbHzrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "df_hs[\"mut\"].hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That is a nice histogram! Okay, we see that some values are less frequent than others, but again, there is no distribution better representing these data than uniform, so the verdict is: uniform-ish."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many unique pairs and triplets are there?\n",
    "\n",
    "Okay, so what do we know so far?\n",
    "\n",
    "- There are 23 different chromosomes\n",
    "- There are 1063 different positions\n",
    "- There are 16 different mutations\n",
    "\n",
    "Now, if these values are independent of each other then we have 23 * 1063 * 16 = 391.184 unique triplets! That is NOT GOOD! If this is true, then we have two options:\n",
    "\n",
    "- One-hot encoding for existance of every possible triplets in the file. Now, average healthy person file contains 486 rows and average sick person files contains 483 rows. Let's say on average every one of them has 485 rows to make calculations easier. It would mean, that this one-hot encoded vector contains only around 0.12% of ones and rest is zeros. That is VERY NOT GOOD. Such inbalance of values may prove to be very difficult for classifier to learn.\n",
    "- Embedding layers. Ach, yes. Embedding layers! Magical layers allowing us to treat symbolic features like computable vectors. They are wonderful featues, but... if you can not use them, don't use them. They are a bit specific and complicated, when the taks seems to be quite easy. It's just a binary classification after all! Also, if we use embedding layers, then out of the blue we have to ensure that triplets will appear in the same order, so each triplet would meet weights of FC layers designed to train them. It would probably mean some complicated neural network architecture like Time-Delay Neural Network, Convolutional Neural Network or even LSTM Network! Again, it is a binary classification task after all!\n",
    "\n",
    "So before we start to panic about difficultiness of the task, let's see if chromosomes, positions and mutations are truly independent from each other."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>mut</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chrX</th>\n",
       "      <th>842</th>\n",
       "      <td>92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY</th>\n",
       "      <th>1149</th>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr13</th>\n",
       "      <th>1407</th>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrX</th>\n",
       "      <th>1767</th>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2</th>\n",
       "      <th>308</th>\n",
       "      <td>84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr11</th>\n",
       "      <th>1613</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21</th>\n",
       "      <th>784</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7</th>\n",
       "      <th>1166</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">chr21</th>\n",
       "      <th>536</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1407 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            mut\n",
       "chr   pos      \n",
       "chrX  842    92\n",
       "chrY  1149   91\n",
       "chr13 1407   86\n",
       "chrX  1767   85\n",
       "chr2  308    84\n",
       "...         ...\n",
       "chr11 1613   11\n",
       "chr21 784    11\n",
       "chr7  1166   10\n",
       "chr21 536    10\n",
       "      1733    9\n",
       "\n",
       "[1407 rows x 1 columns]"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_pos = df_hs.groupby([\"chr\", \"pos\"])\n",
    "chr_pos.count().sort_values(\"mut\", ascending=[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not so bad! There are only 1407 unique pairs of chr and pos instead of 23 * 1063 = 24.449! Moreover, it seems that there are more probable pairs (chrX + 842 with frequency of 92) than others (chr21 + 536). Let's check more!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>mut</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr16</th>\n",
       "      <th>A/G</th>\n",
       "      <td>393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr2</th>\n",
       "      <th>C/C</th>\n",
       "      <td>342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">chr7</th>\n",
       "      <th>G/T</th>\n",
       "      <td>342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C/G</th>\n",
       "      <td>321</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr6</th>\n",
       "      <th>C/C</th>\n",
       "      <td>319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr9</th>\n",
       "      <th>C/T</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY</th>\n",
       "      <th>C/C</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3</th>\n",
       "      <th>A/T</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21</th>\n",
       "      <th>C/T</th>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr12</th>\n",
       "      <th>C/A</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>361 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos\n",
       "chr   mut     \n",
       "chr16 A/G  393\n",
       "chr2  C/C  342\n",
       "chr7  G/T  342\n",
       "      C/G  321\n",
       "chr6  C/C  319\n",
       "...        ...\n",
       "chr9  C/T   19\n",
       "chrY  C/C   18\n",
       "chr3  A/T   18\n",
       "chr21 C/T   16\n",
       "chr12 C/A   14\n",
       "\n",
       "[361 rows x 1 columns]"
      ]
     },
     "execution_count": 241,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr_mut = df_hs.groupby([\"chr\", \"mut\"])\n",
    "chr_mut.count().sort_values(\"pos\", ascending=[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well... not so good here. 361 vs. 23 * 16 = 368"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 242,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>509</th>\n",
       "      <th>T/T</th>\n",
       "      <td>94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1794</th>\n",
       "      <th>C/T</th>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360</th>\n",
       "      <th>C/T</th>\n",
       "      <td>91</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1203</th>\n",
       "      <th>A/G</th>\n",
       "      <td>87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>609</th>\n",
       "      <th>A/T</th>\n",
       "      <td>82</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>784</th>\n",
       "      <th>C/A</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <th>C/T</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1166</th>\n",
       "      <th>C/G</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <th>G/G</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733</th>\n",
       "      <th>A/C</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1407 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          chr\n",
       "pos  mut     \n",
       "509  T/T   94\n",
       "1794 C/T   93\n",
       "1360 C/T   91\n",
       "1203 A/G   87\n",
       "609  A/T   82\n",
       "...       ...\n",
       "784  C/A   11\n",
       "62   C/T   11\n",
       "1166 C/G   10\n",
       "536  G/G   10\n",
       "1733 A/C    9\n",
       "\n",
       "[1407 rows x 1 columns]"
      ]
     },
     "execution_count": 242,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_mut = df_hs.groupby([\"pos\", \"mut\"])\n",
    "pos_mut.count().sort_values(\"chr\", ascending=[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1407 vs. 1063 * 16 = 17.008. Not bad! Now, let's check how many unique triplets are there."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 243,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1434"
      ]
     },
     "execution_count": 243,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique = df_hs.groupby([\"chr\", \"pos\", \"mut\"])\n",
    "len(unique)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wow!... Only 1434 unique triplets instead of over 300 thousands! Not bad. Not bad at all! Now... we need a histogram of they apperance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr18</th>\n",
       "      <th>1597</th>\n",
       "      <th>G/A</th>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">chr8</th>\n",
       "      <th>1204</th>\n",
       "      <th>T/C</th>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1712</th>\n",
       "      <th>C/G</th>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr16</th>\n",
       "      <th>1685</th>\n",
       "      <th>A/G</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr10</th>\n",
       "      <th>609</th>\n",
       "      <th>A/T</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr17</th>\n",
       "      <th>62</th>\n",
       "      <th>C/T</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">chr21</th>\n",
       "      <th>784</th>\n",
       "      <th>C/A</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>536</th>\n",
       "      <th>G/G</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr7</th>\n",
       "      <th>1166</th>\n",
       "      <th>C/G</th>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr21</th>\n",
       "      <th>1733</th>\n",
       "      <th>A/C</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1434 rows  1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                freq\n",
       "chr   pos  mut      \n",
       "chr18 1597 G/A    62\n",
       "chr8  1204 T/C    56\n",
       "      1712 C/G    55\n",
       "chr16 1685 A/G    54\n",
       "chr10 609  A/T    53\n",
       "...              ...\n",
       "chr17 62   C/T    11\n",
       "chr21 784  C/A    11\n",
       "      536  G/G    10\n",
       "chr7  1166 C/G    10\n",
       "chr21 1733 A/C     9\n",
       "\n",
       "[1434 rows x 1 columns]"
      ]
     },
     "execution_count": 247,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_hs[\"freq\"] = 0\n",
    "unique = df_hs.groupby([\"chr\", \"pos\", \"mut\"])\n",
    "unique = unique.count().sort_values(\"freq\", ascending=[0])\n",
    "unique"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not bad, not bad. Now it is time to get some statistics on how often these triplets appear"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average: 34.49511854951186, Min: 9, Max: 62, Std: 9.415944881553063\n"
     ]
    }
   ],
   "source": [
    "freqs = unique[\"freq\"].to_numpy()\n",
    "\n",
    "print(f\"Average: {freqs.mean()}, Min: {freqs.min()}, Max: {freqs.max()}, Std: {freqs.std()}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How about quickly plotting a histogram?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 249,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([ 29.,  88., 141., 188., 237., 310., 327., 104.,   9.,   1.]),\n",
       " array([ 9. , 14.3, 19.6, 24.9, 30.2, 35.5, 40.8, 46.1, 51.4, 56.7, 62. ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 249,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAD4CAYAAAAXUaZHAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAPpElEQVR4nO3df6zddX3H8edLwB9TI2Dvmq7tdpl2MzUZhdwgRrMgREU0KyaOQDZtDEn9AxNMXLbqP+oyEkymTLONpAqzLio0KqMRojIkcf4heIsIlEqsWEKbQq8C/pgZSfG9P86neiy3vT/OvT3th+cjOTnf7+f7+Z7v+xMOr/vN53y/36aqkCT15QXjLkCStPQMd0nqkOEuSR0y3CWpQ4a7JHXo1HEXALBixYqanJwcdxmSdFLZuXPnT6tqYrZtJ0S4T05OMj09Pe4yJOmkkuTRo21zWkaSOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjp0QtyhKum5JrfcNpbj7r327WM5rpaWZ+6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQh71CV5jCuO0WlUXjmLkkdMtwlqUOGuyR1aM5wT/LiJPck+UGSXUk+1trPSnJ3kj1Jbk7ywtb+ora+p22fXOYxSJKOMJ8z92eAC6vqbGADcHGS84GPA9dV1auBp4ArW/8rgada+3WtnyTpOJoz3GvgV231tPYq4ELgy619G3BpW97Y1mnbL0qSpSpYkjS3ec25JzklyX3AQeAO4MfA01V1qHXZB6xuy6uBxwDa9p8Dr5zlMzcnmU4yPTMzM9IgJEm/b17hXlXPVtUGYA1wHvCaUQ9cVVuraqqqpiYmJkb9OEnSkAVdLVNVTwN3Aa8HTk9y+CaoNcD+trwfWAvQtr8C+NlSFCtJmp/5XC0zkeT0tvwS4M3AbgYh/67WbRNwa1ve0dZp279VVbWENUuS5jCfxw+sArYlOYXBH4PtVfW1JA8BNyX5J+D7wA2t/w3AfybZAzwJXL4MdUuSjmHOcK+q+4FzZml/hMH8+5Ht/wf89ZJUJ0laFO9QlaQOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHZoz3JOsTXJXkoeS7EpydWv/aJL9Se5rr0uG9vlQkj1JHk7y1uUcgCTpuU6dR59DwAer6t4kLwd2Jrmjbbuuqv55uHOS9cDlwGuBPwL+O8mfVdWzS1m4JOno5jxzr6oDVXVvW/4lsBtYfYxdNgI3VdUzVfUTYA9w3lIUK0manwXNuSeZBM4B7m5N709yf5Ibk5zR2lYDjw3tto9Z/hgk2ZxkOsn0zMzMwiuXJB3VvMM9ycuArwAfqKpfANcDrwI2AAeATyzkwFW1taqmqmpqYmJiIbtKkuYwr3BPchqDYP9CVX0VoKqeqKpnq+o3wGf43dTLfmDt0O5rWpsk6TiZz9UyAW4AdlfVJ4faVw11eyfwYFveAVye5EVJzgLWAfcsXcmSpLnM52qZNwDvBh5Icl9r+zBwRZINQAF7gfcBVNWuJNuBhxhcaXOVV8pI0vE1Z7hX1XeAzLLp9mPscw1wzQh1SZJG4B2qktQhw12SOmS4S1KH5vODqjR2k1tuG3cJ0knFM3dJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ3OGe5K1Se5K8lCSXUmubu1nJrkjyY/a+xmtPUk+nWRPkvuTnLvcg5Ak/b75nLkfAj5YVeuB84GrkqwHtgB3VtU64M62DvA2YF17bQauX/KqJUnHNGe4V9WBqrq3Lf8S2A2sBjYC21q3bcClbXkj8Pka+C5wepJVS124JOnoFjTnnmQSOAe4G1hZVQfapseBlW15NfDY0G77WtuRn7U5yXSS6ZmZmYXWLUk6hnmHe5KXAV8BPlBVvxjeVlUF1EIOXFVbq2qqqqYmJiYWsqskaQ7zCvckpzEI9i9U1Vdb8xOHp1va+8HWvh9YO7T7mtYmSTpO5nO1TIAbgN1V9cmhTTuATW15E3DrUPt72lUz5wM/H5q+kSQdB6fOo88bgHcDDyS5r7V9GLgW2J7kSuBR4LK27XbgEmAP8GvgvUtZsCRpbnOGe1V9B8hRNl80S/8CrhqxLknSCLxDVZI6ZLhLUofmM+cu/dbkltvGXYKkefDMXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHZoz3JPcmORgkgeH2j6aZH+S+9rrkqFtH0qyJ8nDSd66XIVLko5uPmfunwMunqX9uqra0F63AyRZD1wOvLbt8+9JTlmqYiVJ8zNnuFfVt4En5/l5G4GbquqZqvoJsAc4b4T6JEmLMMqc+/uT3N+mbc5obauBx4b67Gttz5Fkc5LpJNMzMzMjlCFJOtJiw/164FXABuAA8ImFfkBVba2qqaqampiYWGQZkqTZLCrcq+qJqnq2qn4DfIbfTb3sB9YOdV3T2iRJx9Giwj3JqqHVdwKHr6TZAVye5EVJzgLWAfeMVqIkaaFOnatDki8BFwArkuwDPgJckGQDUMBe4H0AVbUryXbgIeAQcFVVPbsslUuSjmrOcK+qK2ZpvuEY/a8BrhmlKEnSaOYMd514JrfcNu4SJJ3gfPyAJHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUofmDPckNyY5mOTBobYzk9yR5Eft/YzWniSfTrInyf1Jzl3O4iVJszt1Hn0+B/wr8Pmhti3AnVV1bZItbf0fgLcB69rrdcD17V3SSWJyy21jO/bea98+tmP3Zs4z96r6NvDkEc0bgW1teRtw6VD752vgu8DpSVYtUa2SpHla7Jz7yqo60JYfB1a25dXAY0P99rW250iyOcl0kumZmZlFliFJms3IP6hWVQG1iP22VtVUVU1NTEyMWoYkachiw/2Jw9Mt7f1ga98PrB3qt6a1SZKOo8WG+w5gU1veBNw61P6edtXM+cDPh6ZvJEnHyZxXyyT5EnABsCLJPuAjwLXA9iRXAo8Cl7XutwOXAHuAXwPvXYaaJUlzmDPcq+qKo2y6aJa+BVw1alGSpNF4h6okdchwl6QOzecOVR3FOO/kk6Rj8cxdkjpkuEtShwx3SeqQ4S5JHTLcJalDhrskdchwl6QOGe6S1CHDXZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDhnuktQhw12SOmS4S1KHDHdJ6pDhLkkdGukfyE6yF/gl8CxwqKqmkpwJ3AxMAnuBy6rqqdHKlCQtxFKcub+pqjZU1VRb3wLcWVXrgDvbuiTpOFqOaZmNwLa2vA24dBmOIUk6hlHDvYBvJtmZZHNrW1lVB9ry48DK2XZMsjnJdJLpmZmZEcuQJA0bac4deGNV7U/yh8AdSX44vLGqKknNtmNVbQW2AkxNTc3aR5K0OCOduVfV/vZ+ELgFOA94IskqgPZ+cNQiJUkLs+hwT/LSJC8/vAy8BXgQ2AFsat02AbeOWqQkaWFGmZZZCdyS5PDnfLGqvp7ke8D2JFcCjwKXjV6mJGkhFh3uVfUIcPYs7T8DLhqlKEnSaLxDVZI6ZLhLUocMd0nqkOEuSR0y3CWpQ4a7JHXIcJekDo36bJmxm9xy27hLkKQTjmfuktQhw12SOmS4S1KHDHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUIcNdkjpkuEtShwx3SerQSf/gMEn9GNeDAPde+/axHHc5eeYuSR0y3CWpQ8sW7kkuTvJwkj1JtizXcSRJz7Us4Z7kFODfgLcB64ErkqxfjmNJkp5ruX5QPQ/YU1WPACS5CdgIPLRMx5OkRRvnv+i2XD/mLle4rwYeG1rfB7xuuEOSzcDmtvqrJA8vcQ0rgJ8u8WeeiJ4P43SM/Xg+jHNBY8zHRzrWnxxtw9guhayqrcDW5fr8JNNVNbVcn3+ieD6M0zH24/kwzhNljMv1g+p+YO3Q+prWJkk6DpYr3L8HrEtyVpIXApcDO5bpWJKkIyzLtExVHUryfuAbwCnAjVW1azmOdQzLNuVzgnk+jNMx9uP5MM4TYoypqnHXIElaYt6hKkkdMtwlqUNdhHuSG5McTPLgUNuZSe5I8qP2fsY4axxVkrVJ7kryUJJdSa5u7d2MM8mLk9yT5AdtjB9r7Wclubs9yuLm9iP9SS/JKUm+n+Rrbb2rcSbZm+SBJPclmW5t3XxfD0tyepIvJ/lhkt1JXn8ijLOLcAc+B1x8RNsW4M6qWgfc2dZPZoeAD1bVeuB84Kr2SIeexvkMcGFVnQ1sAC5Ocj7wceC6qno18BRw5fhKXFJXA7uH1nsc55uqasPQdd89fV8P+xTw9ap6DXA2g/+m4x9nVXXxAiaBB4fWHwZWteVVwMPjrnGJx3sr8OZexwn8AXAvgzubfwqc2tpfD3xj3PUtwfjWMPif/kLga0B6GyewF1hxRFtX31fgFcBPaBennEjj7OXMfTYrq+pAW34cWDnOYpZSkkngHOBuOhtnm6q4DzgI3AH8GHi6qg61LvsYPN7iZPcvwN8Dv2nrr6S/cRbwzSQ72+NGoLPvK3AWMAP8R5ti+2ySl3ICjLPncP+tGvz57OKazyQvA74CfKCqfjG8rYdxVtWzVbWBwZntecBrxlvR0kvyDuBgVe0cdy3L7I1VdS6Dp8NeleQvhzf28H1lcK/QucD1VXUO8L8cMQUzrnH2HO5PJFkF0N4PjrmekSU5jUGwf6GqvtqauxsnQFU9DdzFYHri9CSHb7jr4VEWbwD+Ksle4CYGUzOforNxVtX+9n4QuIXBH+vevq/7gH1VdXdb/zKDsB/7OHsO9x3Apra8icEc9UkrSYAbgN1V9cmhTd2MM8lEktPb8ksY/Kawm0HIv6t1O6nHCFBVH6qqNVU1yeDRHN+qqr+ho3EmeWmSlx9eBt4CPEhH31eAqnoceCzJn7emixg82nzs4+ziDtUkXwIuYPCozSeAjwD/BWwH/hh4FLisqp4cU4kjS/JG4H+AB/jdPO2HGcy7dzHOJH8BbGPwyIoXANur6h+T/CmDM9wzge8Df1tVz4yv0qWT5ALg76rqHT2Ns43llrZ6KvDFqromySvp5Pt6WJINwGeBFwKPAO+lfX8Z4zi7CHdJ0u/reVpGkp63DHdJ6pDhLkkdMtwlqUOGuyR1yHCXpA4Z7pLUof8HwqokWSaNrnsAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay, now we know quite a lot about our data! Let's save for later then:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [],
   "source": [
    "unique.to_csv(\"histogram.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Awesome! Now, let's check how often each of these pairs appears in healthy and sick people"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfh = pd.concat(df_healthy_filtered, axis=0)\n",
    "dfs = pd.concat(df_sick_filtered, axis=0)\n",
    "\n",
    "dfh[\"freq\"] = 0\n",
    "dfh.groupby([\"chr\", \"pos\", \"mut\"]).count().sort_values(\"freq\", ascending=[0]).to_csv(\"healthy_histogram.csv\")\n",
    "\n",
    "dfs[\"freq\"] = 0\n",
    "dfs.groupby([\"chr\", \"pos\", \"mut\"]).count().sort_values(\"freq\", ascending=[0]).to_csv(\"sick_histogram.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Okay! So we have histograms of healthy and sick people. Now let's merge them, to see what triplets are common among all groups, whic are group specific and which are irrelevant for every group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "healthy_hist = pd.read_csv(\"healthy_histogram.csv\")\n",
    "healthy_hist = healthy_hist.rename(columns={\"freq\": \"healthy_freq\"})\n",
    "\n",
    "sick_hist = pd.read_csv(\"sick_histogram.csv\")\n",
    "sick_hist = sick_hist.rename(columns={\"freq\": \"sick_freq\"})\n",
    "\n",
    "merged = healthy_hist.merge(sick_hist, on=[\"chr\", \"pos\", \"mut\"], how=\"outer\")\n",
    "merged[\"freq\"] = merged[\"healthy_freq\"] + merged[\"sick_freq\"]\n",
    "merged.sort_values(\"freq\", ascending=[0])\n",
    "merged.to_csv(\"healthy_sick_histogram.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We finished EDA! So, what did we learned?\n",
    "\n",
    "- There are columns name mismatch between healthy and sick ones. We corrected it\n",
    "- There is an error in data. Rows may contain non-existing chr22. We cleaned it.\n",
    "- Typical file has around 485 triplets, when there are 1434 unique triplets in whole set. It means, that if we encode file using one-hot encoding it will contain around 34% ones. That's awesome! It means that we can use it and our cliassifiers can be simple one.\n",
    "\n",
    "Let's use that knowledge to encode data for a classifier!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Data encoding\n",
    "\n",
    "In this segment we will convert csv files of healthy and sick people so the machine learning classifier will be able to distinguish between healthy and sick people. Thanks to EDA, we know, that we can one-hot encode triplets. Let's get started!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load histogram:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = pd.read_csv(\"histogram.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we can filter out histogram to get rid of values that occur rarely, or maybe to exclude values that are to uniformly spread between healthy and sick alike. For now we won't do it, but it is good to remember about it."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take only these colmuns, that encode gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = hist[[\"chr\", \"pos\", \"mut\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load csvs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load(path: str) -> list:\n",
    "    file_paths = [os.path.join(path, f) for f in os.listdir(path)]\n",
    "    return [pd.read_csv(f) for f in file_paths]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfh = load(\"healthy\")\n",
    "dfs = load(\"sick\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Rename \"CHROMOSOME\" row in sick csvs for \"chr\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfs = [df.rename(columns={\"CHROMOSOME\": \"chr\"}) for df in dfs]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Take only useful rows (\"chr\", \"pos\", \"mut\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfh = [df[[\"chr\", \"pos\", \"mut\"]] for df in dfh]\n",
    "dfs = [df[[\"chr\", \"pos\", \"mut\"]] for df in dfs]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add column \"exists\" = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [],
   "source": [
    "for df in [*dfh, *dfs]:\n",
    "    df[\"exists\"] = 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we want to perform one-hot encoding. It means that, we want to check for every triplet in histogram if such a triplet exists in the file. To speed up this process we will use left-join on histogram and file's dataframe and check \"exist\" column. If triplet exists in file, then \"exist\" column will have 1. If not, then Nan. Let's see how it worked on one frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>exists</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>1597</td>\n",
       "      <td>G/A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>1204</td>\n",
       "      <td>T/C</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>1712</td>\n",
       "      <td>C/G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>1685</td>\n",
       "      <td>A/G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>609</td>\n",
       "      <td>A/T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>chr17</td>\n",
       "      <td>62</td>\n",
       "      <td>C/T</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>chr21</td>\n",
       "      <td>784</td>\n",
       "      <td>C/A</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>chr21</td>\n",
       "      <td>536</td>\n",
       "      <td>G/G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>chr7</td>\n",
       "      <td>1166</td>\n",
       "      <td>C/G</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>chr21</td>\n",
       "      <td>1733</td>\n",
       "      <td>A/C</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1434 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr   pos  mut  exists\n",
       "0     chr18  1597  G/A     NaN\n",
       "1      chr8  1204  T/C     1.0\n",
       "2      chr8  1712  C/G     1.0\n",
       "3     chr16  1685  A/G     NaN\n",
       "4     chr10   609  A/T     1.0\n",
       "...     ...   ...  ...     ...\n",
       "1429  chr17    62  C/T     NaN\n",
       "1430  chr21   784  C/A     NaN\n",
       "1431  chr21   536  G/G     NaN\n",
       "1432   chr7  1166  C/G     NaN\n",
       "1433  chr21  1733  A/C     NaN\n",
       "\n",
       "[1434 rows x 4 columns]"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist.merge(dfh[0], on=[\"chr\", \"pos\", \"mut\"], how=\"left\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not bad! However, Nan means nothing for classifier, so we have to convert them to zeros. Let's check now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>exists</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr18</td>\n",
       "      <td>1597</td>\n",
       "      <td>G/A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr8</td>\n",
       "      <td>1204</td>\n",
       "      <td>T/C</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr8</td>\n",
       "      <td>1712</td>\n",
       "      <td>C/G</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr16</td>\n",
       "      <td>1685</td>\n",
       "      <td>A/G</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr10</td>\n",
       "      <td>609</td>\n",
       "      <td>A/T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>chr17</td>\n",
       "      <td>62</td>\n",
       "      <td>C/T</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>chr21</td>\n",
       "      <td>784</td>\n",
       "      <td>C/A</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>chr21</td>\n",
       "      <td>536</td>\n",
       "      <td>G/G</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>chr7</td>\n",
       "      <td>1166</td>\n",
       "      <td>C/G</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>chr21</td>\n",
       "      <td>1733</td>\n",
       "      <td>A/C</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1434 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        chr   pos  mut  exists\n",
       "0     chr18  1597  G/A     0.0\n",
       "1      chr8  1204  T/C     1.0\n",
       "2      chr8  1712  C/G     1.0\n",
       "3     chr16  1685  A/G     0.0\n",
       "4     chr10   609  A/T     1.0\n",
       "...     ...   ...  ...     ...\n",
       "1429  chr17    62  C/T     0.0\n",
       "1430  chr21   784  C/A     0.0\n",
       "1431  chr21   536  G/G     0.0\n",
       "1432   chr7  1166  C/G     0.0\n",
       "1433  chr21  1733  A/C     0.0\n",
       "\n",
       "[1434 rows x 4 columns]"
      ]
     },
     "execution_count": 274,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist.merge(dfh[0], on=[\"chr\", \"pos\", \"mut\"], how=\"left\").fillna(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Much better! So, let's encode them!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [],
   "source": [
    "def encode(hist: pd.DataFrame, df: pd.DataFrame):\n",
    "    return hist.merge(df, on=[\"chr\", \"pos\", \"mut\"], how=\"left\").fillna(0)[\"exists\"].to_numpy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfh_encoded = [encode(hist,df) for df in dfh]\n",
    "dfs_encoded = [encode(hist,df) for df in dfs]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wonderful! We have collections of numpy arrays containing one-hot encoded vectors for each file. It is time to create lables for each of this vector. We detect Neurofibromatosis, so if someone is sick, then he or she should have label 1 and if not then 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfh_labels = np.zeros(len(dfh_encoded))\n",
    "dfs_labels = np.ones(len(dfs_encoded))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Great! Now, classifiers, which we will use wants to be provided two matrixies: one containing data (one-hot encoded vectors) and one containing labels. It means that, we have to transofrm list of one-hot encoded vectors into one big matrix. Let's go:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We will want to concatenate these vectors by axis = 0, but first we have turn them from 1D vectors of shape 1434 to 2D vectors of shape (1, 1434), so numpy will be able to concat them by first axis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "vectors = [*dfh_encoded, *dfs_encoded]\n",
    "vectors = [v.reshape(1,-1) for v in vectors]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Time to concatenate them!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(102, 1434)"
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X = np.concatenate(vectors, axis=0)\n",
    "X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Outstanding! Now, concatenate labels:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(102,)"
      ]
     },
     "execution_count": 337,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y = np.concatenate([dfh_labels, dfs_labels], axis=0)\n",
    "Y.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Almost done! There is only one more thing to do: shuffle them. Why? Right now there are 60 healthy people in the row and after that there are 42 sick people. It may cause classifier to overfit, so we want to shuffle them to prevent it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [],
   "source": [
    "indicies = np.arange(labels.shape[0])\n",
    "np.random.shuffle(indicies)\n",
    "X = X[indicies, :]\n",
    "Y = Y[indicies]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hooray! We have matricies encoded and shuffled! One last thing to do is to divide X, Y into train, validation and test datasets. Let's say: 60% of data for training, 20% for validation and 20% for testing. Let's go:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_stop = int(Y.shape[0] * 0.6)\n",
    "val_stop = train_stop + int(Y.shape[0] * 0.2)\n",
    "\n",
    "train_x = X[:train_stop, :]\n",
    "val_x = X[train_stop: val_stop, :]\n",
    "test_x = X[val_stop:, :]\n",
    "\n",
    "train_y = Y[:train_stop]\n",
    "val_y = Y[train_stop: val_stop]\n",
    "test_y = Y[val_stop]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We did it! Data are ready to be filled into the classifiers. Let's go to the third step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Neurofibromatosis detection\n",
    "\n",
    "We have performed *Exploratory Data Analysis* and encoded data based on its results. Now it is time to get down with the sickness (Ow-aaaaa!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is a list of classifiers, we will try to classify patients:\n",
    "\n",
    "- SVM (*Support Vector Machine*)\n",
    "- Random Forest\n",
    "- XGBoost\n",
    "\n",
    "We will train classifiers on training data and evaluate them on validation data. The best one will be then tested on test data as a final result of our studies. For evaluation we will use two commonly used metrics: accuracy and F1. Accuracy depicts how often model was right. F1 score on the other hand is harmonic average between precision and recall. Without diving any deeper in the subject let's just say, that F1 is good metric to determine if model does not favorize one class over another.\n",
    "\n",
    "Let's get started!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 356,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evaluate(model):\n",
    "    model.fit(train_x, train_y)\n",
    "    train_pred = model.predict(train_x)\n",
    "    \n",
    "    train_acc = accuracy_score(train_y, train_pred)\n",
    "    train_f1 = f1_score(train_y, train_pred, average=None)\n",
    "    \n",
    "    print(f\"Training: acc: {train_acc}, f1: {train_f1}\")\n",
    "    \n",
    "    val_pred = model.predict(val_x)\n",
    "    \n",
    "    val_acc = accuracy_score(val_y, val_pred)\n",
    "    val_f1 = f1_score(val_y, val_pred, average=None)\n",
    "    \n",
    "    print(f\"Validation: acc: {val_acc}, f1: {val_f1}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 357,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training: acc: 1.0, f1: [1. 1.]\n",
      "Validation: acc: 0.65, f1: [0.78787879 0.        ]\n"
     ]
    }
   ],
   "source": [
    "# Using SVM\n",
    "\n",
    "evaluate(SVC())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 358,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training: acc: 1.0, f1: [1. 1.]\n",
      "Validation: acc: 0.65, f1: [0.77419355 0.22222222]\n"
     ]
    }
   ],
   "source": [
    "# Using Random Forest\n",
    "\n",
    "evaluate(RandomForestClassifier())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training: acc: 1.0, f1: [1. 1.]\n",
      "Validation: acc: 0.65, f1: [0.72       0.53333333]\n"
     ]
    }
   ],
   "source": [
    "# Using XGBoost\n",
    "\n",
    "evaluate(xgboost.XGBClassifier())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not good... Well, that is not suprising, to be honest. There are barely 60 samples of data. This is very, very small amount. Unfortunately, we won't get more, so let's try to accually do seomthing with it. Altghough every classifier got 65% accuracy, XGBoost scored the best f1 score on two classes combined, so we will take XGBoost for tuning and see, if we will be able to produce better results."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. Automatic classifier tuning with Optuna library\n",
    "\n",
    "In this point we will try to fine-tune xgboost using Optuna library for automatic optimization. We will try to find better results, by modifying these parameters:\n",
    "\n",
    "- maximum tree depth\n",
    "- number of estimators\n",
    "- amount of rows to be seen by one tree\n",
    "- regularization lambda parameter\n",
    "- cut-off parameter\n",
    "\n",
    "Let's get started"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 386,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _objective(trial) -> float:\n",
    "    max_depth = trial.suggest_int(\"max_depth\",low=3,high=12)\n",
    "    n_estimators = trial.suggest_int(\"n_estimators\",10,500)\n",
    "    reg_lambda = trial.suggest_loguniform(\"reg_labda\", 0.1, 10.0)\n",
    "    subsample = trial.suggest_uniform(\"subsample\", 0.1, 1.0)\n",
    "    p = trial.suggest_uniform(\"cut-off parameter\", 0.1, 0.9)\n",
    "    xgb = xgboost.XGBClassifier(max_depth = max_depth, n_estimators=n_estimators, \n",
    "                               reg_lambda=reg_lambda, subsample=subsample)\n",
    "    xgb = xgboost.XGBClassifier()\n",
    "    \n",
    "    xgb.fit(train_x, train_y)\n",
    "#     pred = xgb.predict_proba(val_x)\n",
    "#     pred = softmax(pred)[:,1]\n",
    "#     pred = pred > p\n",
    "#     pred = pred.astype(int)\n",
    "    \n",
    "    f1 = min(f1_score(val_y, xgb.predict(val_x), average=None))\n",
    "    return f1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.remove(\"optuna.db\")\n",
    "study = optuna.create_study(study_name=\"xgboost for neurofibromatosis classification test\",\n",
    "                            storage=\"sqlite:///optuna.db\",\n",
    "                            load_if_exists=False)\n",
    "study.optimize(_objective, n_trials=1000)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well... nothing helps. Overfitting is too strong. So, if we are not able to get more data or fine-tune model, then we should try remove some of the featues"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 5. Back to EDA\n",
    "\n",
    "Let's try taking less features in order to prevent such strong overfitting. To do it, we have to analyse healthy_sick_histogram."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 390,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>healthy_freq</th>\n",
       "      <th>sick_freq</th>\n",
       "      <th>freq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>118</td>\n",
       "      <td>chr8</td>\n",
       "      <td>1204</td>\n",
       "      <td>T/C</td>\n",
       "      <td>30</td>\n",
       "      <td>26</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>575</th>\n",
       "      <td>575</td>\n",
       "      <td>chr20</td>\n",
       "      <td>1292</td>\n",
       "      <td>A/C</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>597</th>\n",
       "      <td>597</td>\n",
       "      <td>chr5</td>\n",
       "      <td>668</td>\n",
       "      <td>A/T</td>\n",
       "      <td>24</td>\n",
       "      <td>24</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>69</td>\n",
       "      <td>chr8</td>\n",
       "      <td>1712</td>\n",
       "      <td>C/G</td>\n",
       "      <td>31</td>\n",
       "      <td>24</td>\n",
       "      <td>55</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>947</th>\n",
       "      <td>947</td>\n",
       "      <td>chr2</td>\n",
       "      <td>1848</td>\n",
       "      <td>C/A</td>\n",
       "      <td>17</td>\n",
       "      <td>23</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1106</th>\n",
       "      <td>1106</td>\n",
       "      <td>chrY</td>\n",
       "      <td>372</td>\n",
       "      <td>G/C</td>\n",
       "      <td>14</td>\n",
       "      <td>6</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1360</th>\n",
       "      <td>1360</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1446</td>\n",
       "      <td>C/A</td>\n",
       "      <td>7</td>\n",
       "      <td>6</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1077</th>\n",
       "      <td>1077</td>\n",
       "      <td>chr6</td>\n",
       "      <td>1612</td>\n",
       "      <td>A/A</td>\n",
       "      <td>14</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1416</th>\n",
       "      <td>1416</td>\n",
       "      <td>chr21</td>\n",
       "      <td>1733</td>\n",
       "      <td>A/C</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>542</td>\n",
       "      <td>chr12</td>\n",
       "      <td>508</td>\n",
       "      <td>T/G</td>\n",
       "      <td>24</td>\n",
       "      <td>5</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1434 rows  7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0    chr   pos  mut  healthy_freq  sick_freq  freq\n",
       "118          118   chr8  1204  T/C            30         26    56\n",
       "575          575  chr20  1292  A/C            24         24    48\n",
       "597          597   chr5   668  A/T            24         24    48\n",
       "69            69   chr8  1712  C/G            31         24    55\n",
       "947          947   chr2  1848  C/A            17         23    40\n",
       "...          ...    ...   ...  ...           ...        ...   ...\n",
       "1106        1106   chrY   372  G/C            14          6    20\n",
       "1360        1360   chr6  1446  C/A             7          6    13\n",
       "1077        1077   chr6  1612  A/A            14          5    19\n",
       "1416        1416  chr21  1733  A/C             4          5     9\n",
       "542          542  chr12   508  T/G            24          5    29\n",
       "\n",
       "[1434 rows x 7 columns]"
      ]
     },
     "execution_count": 390,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist = pd.read_csv(\"healthy_sick_histogram.csv\")\n",
    "hist.sort_values(\"sick_freq\", ascending=[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's try adding column which depicts difference between sick and healthy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 391,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist[\"diff\"] = hist[\"sick_freq\"] / hist[\"healthy_freq\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 392,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>chr</th>\n",
       "      <th>pos</th>\n",
       "      <th>mut</th>\n",
       "      <th>healthy_freq</th>\n",
       "      <th>sick_freq</th>\n",
       "      <th>freq</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>1433</td>\n",
       "      <td>chr17</td>\n",
       "      <td>62</td>\n",
       "      <td>C/T</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "      <td>10.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>1430</td>\n",
       "      <td>chr1</td>\n",
       "      <td>570</td>\n",
       "      <td>A/T</td>\n",
       "      <td>3</td>\n",
       "      <td>14</td>\n",
       "      <td>17</td>\n",
       "      <td>4.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>1432</td>\n",
       "      <td>chr16</td>\n",
       "      <td>72</td>\n",
       "      <td>C/T</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>4.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1428</th>\n",
       "      <td>1428</td>\n",
       "      <td>chr14</td>\n",
       "      <td>341</td>\n",
       "      <td>T/G</td>\n",
       "      <td>3</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>4.333333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1427</th>\n",
       "      <td>1427</td>\n",
       "      <td>chr18</td>\n",
       "      <td>627</td>\n",
       "      <td>C/T</td>\n",
       "      <td>3</td>\n",
       "      <td>12</td>\n",
       "      <td>15</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255</th>\n",
       "      <td>255</td>\n",
       "      <td>chr5</td>\n",
       "      <td>1985</td>\n",
       "      <td>C/T</td>\n",
       "      <td>28</td>\n",
       "      <td>8</td>\n",
       "      <td>36</td>\n",
       "      <td>0.285714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>28</td>\n",
       "      <td>chr15</td>\n",
       "      <td>1987</td>\n",
       "      <td>T/A</td>\n",
       "      <td>33</td>\n",
       "      <td>9</td>\n",
       "      <td>42</td>\n",
       "      <td>0.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>512</th>\n",
       "      <td>512</td>\n",
       "      <td>chr9</td>\n",
       "      <td>1527</td>\n",
       "      <td>T/C</td>\n",
       "      <td>24</td>\n",
       "      <td>6</td>\n",
       "      <td>30</td>\n",
       "      <td>0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>131</th>\n",
       "      <td>131</td>\n",
       "      <td>chr1</td>\n",
       "      <td>1298</td>\n",
       "      <td>A/A</td>\n",
       "      <td>30</td>\n",
       "      <td>7</td>\n",
       "      <td>37</td>\n",
       "      <td>0.233333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>542</th>\n",
       "      <td>542</td>\n",
       "      <td>chr12</td>\n",
       "      <td>508</td>\n",
       "      <td>T/G</td>\n",
       "      <td>24</td>\n",
       "      <td>5</td>\n",
       "      <td>29</td>\n",
       "      <td>0.208333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1434 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unnamed: 0    chr   pos  mut  healthy_freq  sick_freq  freq       diff\n",
       "1433        1433  chr17    62  C/T             1         10    11  10.000000\n",
       "1430        1430   chr1   570  A/T             3         14    17   4.666667\n",
       "1432        1432  chr16    72  C/T             3         13    16   4.333333\n",
       "1428        1428  chr14   341  T/G             3         13    16   4.333333\n",
       "1427        1427  chr18   627  C/T             3         12    15   4.000000\n",
       "...          ...    ...   ...  ...           ...        ...   ...        ...\n",
       "255          255   chr5  1985  C/T            28          8    36   0.285714\n",
       "28            28  chr15  1987  T/A            33          9    42   0.272727\n",
       "512          512   chr9  1527  T/C            24          6    30   0.250000\n",
       "131          131   chr1  1298  A/A            30          7    37   0.233333\n",
       "542          542  chr12   508  T/G            24          5    29   0.208333\n",
       "\n",
       "[1434 rows x 8 columns]"
      ]
     },
     "execution_count": 392,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist.sort_values(\"diff\", ascending=[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Ok then... how many are there genes, that are more than 2 times more likely in sick group instead of healthy ones?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35, 8)"
      ]
     },
     "execution_count": 393,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist[hist[\"diff\"] > 2].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How about 1.5?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 394,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(108, 8)"
      ]
     },
     "execution_count": 394,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hist[hist[\"diff\"] > 1.5].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not bad. Let's try to use such hist to encode data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 395,
   "metadata": {},
   "outputs": [],
   "source": [
    "hist = hist[hist[\"diff\"] > 1.5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 396,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfh = load(\"healthy\")\n",
    "dfs = load(\"sick\")\n",
    "dfs = [df.rename(columns={\"CHROMOSOME\": \"chr\"}) for df in dfs]\n",
    "dfh = [df[[\"chr\", \"pos\", \"mut\"]] for df in dfh]\n",
    "dfs = [df[[\"chr\", \"pos\", \"mut\"]] for df in dfs]\n",
    "for df in [*dfh, *dfs]:\n",
    "    df[\"exists\"] = 1\n",
    "dfh_encoded = [encode(hist,df) for df in dfh]\n",
    "dfs_encoded = [encode(hist,df) for df in dfs]\n",
    "dfh_labels = np.zeros(len(dfh_encoded))\n",
    "dfs_labels = np.ones(len(dfs_encoded))\n",
    "vectors = [*dfh_encoded, *dfs_encoded]\n",
    "vectors = [v.reshape(1,-1) for v in vectors]\n",
    "X = np.concatenate(vectors, axis=0)\n",
    "Y = np.concatenate([dfh_labels, dfs_labels], axis=0)\n",
    "indicies = np.arange(labels.shape[0])\n",
    "np.random.shuffle(indicies)\n",
    "X = X[indicies, :]\n",
    "Y = Y[indicies]\n",
    "train_stop = int(Y.shape[0] * 0.6)\n",
    "val_stop = train_stop + int(Y.shape[0] * 0.2)\n",
    "\n",
    "train_x = X[:train_stop, :]\n",
    "val_x = X[train_stop: val_stop, :]\n",
    "test_x = X[val_stop:, :]\n",
    "\n",
    "train_y = Y[:train_stop]\n",
    "val_y = Y[train_stop: val_stop]\n",
    "test_y = Y[val_stop]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 397,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training: acc: 1.0, f1: [1. 1.]\n",
      "Validation: acc: 1.0, f1: [1. 1.]\n"
     ]
    }
   ],
   "source": [
    "# Using SVM\n",
    "\n",
    "evaluate(SVC())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wow! The simplest of classifiers was able to have PERFECT performance. What with the rest of classifiers?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 398,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training: acc: 1.0, f1: [1. 1.]\n",
      "Validation: acc: 0.8, f1: [0.85714286 0.66666667]\n"
     ]
    }
   ],
   "source": [
    "# Using Random Forest\n",
    "\n",
    "evaluate(RandomForestClassifier())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 399,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Training: acc: 1.0, f1: [1. 1.]\n",
      "Validation: acc: 0.7, f1: [0.8 0.4]\n"
     ]
    }
   ],
   "source": [
    "# Using XGBoost\n",
    "\n",
    "evaluate(xgboost.XGBClassifier())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's suprising to be honest, but only slightly. You see... there are three rules of thumb to consider, when we deal with overfitting:\n",
    "\n",
    "- increase number of data acquired\n",
    "- simplify the model\n",
    "- reduce amount of features\n",
    "\n",
    "In other words, make it harder for model to \"learn by heart\" training dataset. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "That's all for today. Have a nice day :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
